{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jaehee831/protein-folding-analysis-bioinformatics/blob/main/rf/examples/diffusion.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tSgCPxIZ1T_A"
      },
      "source": [
        "#**RFdiffusion v1.1.1**\n",
        "RFdiffusion is a method for structure generation, with or without conditional information (a motif, target etc). It can perform a whole range of protein design challenges as we have outlined in the RFdiffusion [manuscript](https://www.biorxiv.org/content/10.1101/2022.12.09.519842v2).\n",
        "\n",
        "\n",
        "For **instructions**, see end of Notebook.\n",
        "\n",
        "**<font color=\"red\">NOTE:</font>**  This is tagged v1.1.1 of the notebook, this notebook may break in the future when colab updates. For latest version see [main](https://colab.research.google.com/github/sokrypton/ColabDesign/blob/main/rf/examples/diffusion.ipynb) branch.\n",
        "\n",
        "Additional Notebooks:\n",
        "\n",
        "- See [diffusion_foldcond](https://colab.research.google.com/github/sokrypton/ColabDesign/blob/v1.1.1/rf/examples/diffusion_foldcond.ipynb) for fold conditioning functionality.\n",
        "\n",
        "- See [original version](https://colab.research.google.com/github/sokrypton/ColabDesign/blob/v1.1.1/rf/examples/diffusion_ori.ipynb) of this notebook (from 31Mar2023).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "cellView": "form",
        "id": "pZQnHLuDCsZm",
        "outputId": "28769b25-19c7-4bbc-8a2e-3955a154a485",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "installing RFdiffusion...\n",
            "installing ColabDesign...\n",
            "downloading RFdiffusion params...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/content/RFdiffusion/Track_module.py:241: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.\n",
            "  @torch.cuda.amp.autocast(enabled=False)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 9.54 s, sys: 1.44 s, total: 11 s\n",
            "Wall time: 2min 45s\n"
          ]
        }
      ],
      "source": [
        "#@title setup **RFdiffusion** (~3min)\n",
        "%%time\n",
        "import os, time, signal\n",
        "import sys, random, string, re\n",
        "if not os.path.isdir(\"params\"):\n",
        "  os.system(\"apt-get install aria2\")\n",
        "  os.system(\"mkdir params\")\n",
        "  # send param download into background\n",
        "  os.system(\"(\\\n",
        "  aria2c -q -x 16 https://files.ipd.uw.edu/krypton/schedules.zip; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/6f5902ac237024bdd0c176cb93063dc4/Base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/e29311f6f1bf1af907f9ef9f44b8328b/Complex_base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 https://storage.googleapis.com/alphafold/alphafold_params_2022-12-06.tar; \\\n",
        "  tar -xf alphafold_params_2022-12-06.tar -C params; \\\n",
        "  touch params/done.txt) &\")\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion\"):\n",
        "  print(\"installing RFdiffusion...\")\n",
        "  os.system(\"git clone https://github.com/sokrypton/RFdiffusion.git\")\n",
        "  os.system(\"pip install jedi omegaconf hydra-core icecream pyrsistent pynvml decorator\")\n",
        "  os.system(\"pip install git+https://github.com/NVIDIA/dllogger#egg=dllogger\")\n",
        "  # 17Mar2024: adding --no-dependencies to avoid installing nvidia-cuda-* dependencies\n",
        "  os.system(\"pip install --no-dependencies dgl==2.0.0 -f https://data.dgl.ai/wheels/cu121/repo.html\")\n",
        "  os.system(\"pip install --no-dependencies e3nn==0.5.5 opt_einsum_fx\")\n",
        "  os.system(\"cd RFdiffusion/env/SE3Transformer; pip install .\")\n",
        "  os.system(\"wget -qnc https://files.ipd.uw.edu/krypton/ananas\")\n",
        "  os.system(\"chmod +x ananas\")\n",
        "\n",
        "if not os.path.isdir(\"colabdesign\"):\n",
        "  print(\"installing ColabDesign...\")\n",
        "  os.system(\"pip -q install git+https://github.com/sokrypton/ColabDesign.git@v1.1.1\")\n",
        "  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabdesign colabdesign\")\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion/models\"):\n",
        "  print(\"downloading RFdiffusion params...\")\n",
        "  os.system(\"mkdir RFdiffusion/models\")\n",
        "  models = [\"Base_ckpt.pt\",\"Complex_base_ckpt.pt\"]\n",
        "  for m in models:\n",
        "    while os.path.isfile(f\"{m}.aria2\"):\n",
        "      time.sleep(5)\n",
        "  os.system(f\"mv {' '.join(models)} RFdiffusion/models\")\n",
        "  os.system(\"unzip schedules.zip; rm schedules.zip\")\n",
        "\n",
        "if 'RFdiffusion' not in sys.path:\n",
        "  os.environ[\"DGLBACKEND\"] = \"pytorch\"\n",
        "  sys.path.append('RFdiffusion')\n",
        "\n",
        "from google.colab import files\n",
        "import json\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from IPython.display import display, HTML\n",
        "import ipywidgets as widgets\n",
        "import py3Dmol\n",
        "\n",
        "from inference.utils import parse_pdb\n",
        "from colabdesign.rf.utils import get_ca\n",
        "from colabdesign.rf.utils import fix_contigs, fix_partial_contigs, fix_pdb, sym_it\n",
        "from colabdesign.shared.protein import pdb_to_string\n",
        "from colabdesign.shared.plot import plot_pseudo_3D\n",
        "\n",
        "def get_pdb(pdb_code=None):\n",
        "  if pdb_code is None or pdb_code == \"\":\n",
        "    upload_dict = files.upload()\n",
        "    pdb_string = upload_dict[list(upload_dict.keys())[0]]\n",
        "    with open(\"tmp.pdb\",\"wb\") as out: out.write(pdb_string)\n",
        "    return \"tmp.pdb\"\n",
        "  elif os.path.isfile(pdb_code):\n",
        "    return pdb_code\n",
        "  elif len(pdb_code) == 4:\n",
        "    if not os.path.isfile(f\"{pdb_code}.pdb1\"):\n",
        "      os.system(f\"wget -qnc https://files.rcsb.org/download/{pdb_code}.pdb1.gz\")\n",
        "      os.system(f\"gunzip {pdb_code}.pdb1.gz\")\n",
        "    return f\"{pdb_code}.pdb1\"\n",
        "  else:\n",
        "    os.system(f\"wget -qnc https://alphafold.ebi.ac.uk/files/AF-{pdb_code}-F1-model_v3.pdb\")\n",
        "    return f\"AF-{pdb_code}-F1-model_v3.pdb\"\n",
        "\n",
        "def run_ananas(pdb_str, path, sym=None):\n",
        "  pdb_filename = f\"outputs/{path}/ananas_input.pdb\"\n",
        "  out_filename = f\"outputs/{path}/ananas.json\"\n",
        "  with open(pdb_filename,\"w\") as handle:\n",
        "    handle.write(pdb_str)\n",
        "\n",
        "  cmd = f\"./ananas {pdb_filename} -u -j {out_filename}\"\n",
        "  if sym is None: os.system(cmd)\n",
        "  else: os.system(f\"{cmd} {sym}\")\n",
        "\n",
        "  # parse results\n",
        "  try:\n",
        "    out = json.loads(open(out_filename,\"r\").read())\n",
        "    results,AU = out[0], out[-1][\"AU\"]\n",
        "    group = AU[\"group\"]\n",
        "    chains = AU[\"chain names\"]\n",
        "    rmsd = results[\"Average_RMSD\"]\n",
        "    print(f\"AnAnaS detected {group} symmetry at RMSD:{rmsd:.3}\")\n",
        "\n",
        "    C = np.array(results['transforms'][0]['CENTER'])\n",
        "    A = [np.array(t[\"AXIS\"]) for t in results['transforms']]\n",
        "\n",
        "    # apply symmetry and filter to the asymmetric unit\n",
        "    new_lines = []\n",
        "    for line in pdb_str.split(\"\\n\"):\n",
        "      if line.startswith(\"ATOM\"):\n",
        "        chain = line[21:22]\n",
        "        if chain in chains:\n",
        "          x = np.array([float(line[i:(i+8)]) for i in [30,38,46]])\n",
        "          if group[0] == \"c\":\n",
        "            x = sym_it(x,C,A[0])\n",
        "          if group[0] == \"d\":\n",
        "            x = sym_it(x,C,A[1],A[0])\n",
        "          coord_str = \"\".join([\"{:8.3f}\".format(a) for a in x])\n",
        "          new_lines.append(line[:30]+coord_str+line[54:])\n",
        "      else:\n",
        "        new_lines.append(line)\n",
        "    return results, \"\\n\".join(new_lines)\n",
        "\n",
        "  except:\n",
        "    return None, pdb_str\n",
        "\n",
        "def run(command, steps, num_designs=1, visual=\"none\"):\n",
        "\n",
        "  def run_command_and_get_pid(command):\n",
        "    pid_file = '/dev/shm/pid'\n",
        "    os.system(f'nohup {command} > /dev/null & echo $! > {pid_file}')\n",
        "    with open(pid_file, 'r') as f:\n",
        "      pid = int(f.read().strip())\n",
        "    os.remove(pid_file)\n",
        "    return pid\n",
        "  def is_process_running(pid):\n",
        "    try:\n",
        "      os.kill(pid, 0)\n",
        "    except OSError:\n",
        "      return False\n",
        "    else:\n",
        "      return True\n",
        "\n",
        "  run_output = widgets.Output()\n",
        "  progress = widgets.FloatProgress(min=0, max=1, description='running', bar_style='info')\n",
        "  display(widgets.VBox([progress, run_output]))\n",
        "\n",
        "  # clear previous run\n",
        "  for n in range(steps):\n",
        "    if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
        "      os.remove(f\"/dev/shm/{n}.pdb\")\n",
        "\n",
        "  pid = run_command_and_get_pid(command)\n",
        "  try:\n",
        "    fail = False\n",
        "    for _ in range(num_designs):\n",
        "\n",
        "      # for each step check if output generated\n",
        "      for n in range(steps):\n",
        "        wait = True\n",
        "        while wait and not fail:\n",
        "          time.sleep(0.1)\n",
        "          if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
        "            pdb_str = open(f\"/dev/shm/{n}.pdb\").read()\n",
        "            if pdb_str[-3:] == \"TER\":\n",
        "              wait = False\n",
        "            elif not is_process_running(pid):\n",
        "              fail = True\n",
        "          elif not is_process_running(pid):\n",
        "            fail = True\n",
        "\n",
        "        if fail:\n",
        "          progress.bar_style = 'danger'\n",
        "          progress.description = \"failed\"\n",
        "          break\n",
        "\n",
        "        else:\n",
        "          progress.value = (n+1) / steps\n",
        "          if visual != \"none\":\n",
        "            with run_output:\n",
        "              run_output.clear_output(wait=True)\n",
        "              if visual == \"image\":\n",
        "                xyz, bfact = get_ca(f\"/dev/shm/{n}.pdb\", get_bfact=True)\n",
        "                fig = plt.figure()\n",
        "                fig.set_dpi(100);fig.set_figwidth(6);fig.set_figheight(6)\n",
        "                ax1 = fig.add_subplot(111);ax1.set_xticks([]);ax1.set_yticks([])\n",
        "                plot_pseudo_3D(xyz, c=bfact, cmin=0.5, cmax=0.9, ax=ax1)\n",
        "                plt.show()\n",
        "              if visual == \"interactive\":\n",
        "                view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "                view.addModel(pdb_str,'pdb')\n",
        "                view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':0.5,'max':0.9}}})\n",
        "                view.zoomTo()\n",
        "                view.show()\n",
        "        if os.path.exists(f\"/dev/shm/{n}.pdb\"):\n",
        "          os.remove(f\"/dev/shm/{n}.pdb\")\n",
        "      if fail:\n",
        "        progress.bar_style = 'danger'\n",
        "        progress.description = \"failed\"\n",
        "        break\n",
        "\n",
        "    while is_process_running(pid):\n",
        "      time.sleep(0.1)\n",
        "\n",
        "  except KeyboardInterrupt:\n",
        "    os.kill(pid, signal.SIGTERM)\n",
        "    progress.bar_style = 'danger'\n",
        "    progress.description = \"stopped\"\n",
        "\n",
        "def run_diffusion(contigs, path, pdb=None, iterations=50,\n",
        "                  symmetry=\"none\", order=1, hotspot=None,\n",
        "                  chains=None, add_potential=False,\n",
        "                  num_designs=1, visual=\"none\"):\n",
        "\n",
        "  full_path = f\"outputs/{path}\"\n",
        "  os.makedirs(full_path, exist_ok=True)\n",
        "  opts = [f\"inference.output_prefix={full_path}\",\n",
        "          f\"inference.num_designs={num_designs}\"]\n",
        "\n",
        "  if chains == \"\": chains = None\n",
        "\n",
        "  # determine symmetry type\n",
        "  if symmetry in [\"auto\",\"cyclic\",\"dihedral\"]:\n",
        "    if symmetry == \"auto\":\n",
        "      sym, copies = None, 1\n",
        "    else:\n",
        "      sym, copies = {\"cyclic\":(f\"c{order}\",order),\n",
        "                     \"dihedral\":(f\"d{order}\",order*2)}[symmetry]\n",
        "  else:\n",
        "    symmetry = None\n",
        "    sym, copies = None, 1\n",
        "\n",
        "  # determine mode\n",
        "  contigs = contigs.replace(\",\",\" \").replace(\":\",\" \").split()\n",
        "  is_fixed, is_free = False, False\n",
        "  fixed_chains = []\n",
        "  for contig in contigs:\n",
        "    for x in contig.split(\"/\"):\n",
        "      a = x.split(\"-\")[0]\n",
        "      if a[0].isalpha():\n",
        "        is_fixed = True\n",
        "        if a[0] not in fixed_chains:\n",
        "          fixed_chains.append(a[0])\n",
        "      if a.isnumeric():\n",
        "        is_free = True\n",
        "  if len(contigs) == 0 or not is_free:\n",
        "    mode = \"partial\"\n",
        "  elif is_fixed:\n",
        "    mode = \"fixed\"\n",
        "  else:\n",
        "    mode = \"free\"\n",
        "\n",
        "  # fix input contigs\n",
        "  if mode in [\"partial\",\"fixed\"]:\n",
        "    pdb_str = pdb_to_string(get_pdb(pdb), chains=chains)\n",
        "    if symmetry == \"auto\":\n",
        "      a, pdb_str = run_ananas(pdb_str, path)\n",
        "      if a is None:\n",
        "        print(f'ERROR: no symmetry detected')\n",
        "        symmetry = None\n",
        "        sym, copies = None, 1\n",
        "      else:\n",
        "        if a[\"group\"][0] == \"c\":\n",
        "          symmetry = \"cyclic\"\n",
        "          sym, copies = a[\"group\"], int(a[\"group\"][1:])\n",
        "        elif a[\"group\"][0] == \"d\":\n",
        "          symmetry = \"dihedral\"\n",
        "          sym, copies = a[\"group\"], 2 * int(a[\"group\"][1:])\n",
        "        else:\n",
        "          print(f'ERROR: the detected symmetry ({a[\"group\"]}) not currently supported')\n",
        "          symmetry = None\n",
        "          sym, copies = None, 1\n",
        "\n",
        "    elif mode == \"fixed\":\n",
        "      pdb_str = pdb_to_string(pdb_str, chains=fixed_chains)\n",
        "\n",
        "    pdb_filename = f\"{full_path}/input.pdb\"\n",
        "    with open(pdb_filename, \"w\") as handle:\n",
        "      handle.write(pdb_str)\n",
        "\n",
        "    parsed_pdb = parse_pdb(pdb_filename)\n",
        "    opts.append(f\"inference.input_pdb={pdb_filename}\")\n",
        "    if mode in [\"partial\"]:\n",
        "      iterations = int(80 * (iterations / 200))\n",
        "      opts.append(f\"diffuser.partial_T={iterations}\")\n",
        "      contigs = fix_partial_contigs(contigs, parsed_pdb)\n",
        "    else:\n",
        "      opts.append(f\"diffuser.T={iterations}\")\n",
        "      contigs = fix_contigs(contigs, parsed_pdb)\n",
        "  else:\n",
        "    opts.append(f\"diffuser.T={iterations}\")\n",
        "    parsed_pdb = None\n",
        "    contigs = fix_contigs(contigs, parsed_pdb)\n",
        "\n",
        "  if hotspot is not None and hotspot != \"\":\n",
        "    opts.append(f\"ppi.hotspot_res=[{hotspot}]\")\n",
        "\n",
        "  # setup symmetry\n",
        "  if sym is not None:\n",
        "    sym_opts = [\"--config-name symmetry\", f\"inference.symmetry={sym}\"]\n",
        "    if add_potential:\n",
        "      sym_opts += [\"'potentials.guiding_potentials=[\\\"type:olig_contacts,weight_intra:1,weight_inter:0.1\\\"]'\",\n",
        "                   \"potentials.olig_intra_all=True\",\"potentials.olig_inter_all=True\",\n",
        "                   \"potentials.guide_scale=2\",\"potentials.guide_decay=quadratic\"]\n",
        "    opts = sym_opts + opts\n",
        "    contigs = sum([contigs] * copies,[])\n",
        "\n",
        "  opts.append(f\"'contigmap.contigs=[{' '.join(contigs)}]'\")\n",
        "  opts += [\"inference.dump_pdb=True\",\"inference.dump_pdb_path='/dev/shm'\"]\n",
        "\n",
        "  print(\"mode:\", mode)\n",
        "  print(\"output:\", full_path)\n",
        "  print(\"contigs:\", contigs)\n",
        "\n",
        "  opts_str = \" \".join(opts)\n",
        "  cmd = f\"./RFdiffusion/run_inference.py {opts_str}\"\n",
        "  print(cmd)\n",
        "\n",
        "  # RUN\n",
        "  run(cmd, iterations, num_designs, visual=visual)\n",
        "\n",
        "  # fix pdbs\n",
        "  for n in range(num_designs):\n",
        "    pdbs = [f\"outputs/traj/{path}_{n}_pX0_traj.pdb\",\n",
        "            f\"outputs/traj/{path}_{n}_Xt-1_traj.pdb\",\n",
        "            f\"{full_path}_{n}.pdb\"]\n",
        "    for pdb in pdbs:\n",
        "      with open(pdb,\"r\") as handle: pdb_str = handle.read()\n",
        "      with open(pdb,\"w\") as handle: handle.write(fix_pdb(pdb_str, contigs))\n",
        "\n",
        "  return contigs, copies"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {
        "cellView": "form",
        "id": "TuRUfQJZ4vkM",
        "outputId": "06cad9d7-bafd-4699-c3a2-4ff7dc74fcfd",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 435
        }
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "invalid literal for int() with base 10: '[10'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
            "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36mrun_diffusion\u001b[0;34m(contigs, path, pdb, iterations, symmetry, order, hotspot, chains, add_potential, num_designs, visual)\u001b[0m\n",
            "\u001b[0;32m/content/colabdesign/rf/utils.py\u001b[0m in \u001b[0;36mfix_contigs\u001b[0;34m(contigs, parsed_pdb)\u001b[0m\n\u001b[1;32m    129\u001b[0m         \u001b[0mY\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{x}-{x}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;34m\"/\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mY\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 131\u001b[0;31m   \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mfix_contig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcontigs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mfix_pdb\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpdb_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontigs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/content/colabdesign/rf/utils.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    129\u001b[0m         \u001b[0mY\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{x}-{x}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;34m\"/\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mY\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 131\u001b[0;31m   \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mfix_contig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcontigs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mfix_pdb\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpdb_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontigs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/content/colabdesign/rf/utils.py\u001b[0m in \u001b[0;36mfix_contig\u001b[0;34m(contig)\u001b[0m\n\u001b[1;32m    124\u001b[0m         \u001b[0;31m# sample length\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m         \u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0me\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"-\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 126\u001b[0;31m         \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    127\u001b[0m         \u001b[0mY\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{m}-{m}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    128\u001b[0m       \u001b[0;32melif\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misnumeric\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mx\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"0\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '[10'"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "#@title run **RFdiffusion** to generate a backbone\n",
        "name = \"test\" #@param {type:\"string\"}\n",
        "contigs = \"[10-30/A46-46/5-10/A81-81/5-10/A136-137/20-40]\" #@param {type:\"string\"}\n",
        "pdb = \"6HN3\" #@param {type:\"string\"}\n",
        "iterations = 50 #@param [\"25\", \"50\", \"100\", \"150\", \"200\"] {type:\"raw\"}\n",
        "hotspot = \"A46, A81, A136, A137\" #@param {type:\"string\"}\n",
        "num_designs = 1 #@param [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\"] {type:\"raw\"}\n",
        "visual = \"image\" #@param [\"none\", \"image\", \"interactive\"]\n",
        "#@markdown ---\n",
        "#@markdown **symmetry** settings\n",
        "#@markdown ---\n",
        "symmetry = \"none\" #@param [\"none\", \"auto\", \"cyclic\", \"dihedral\"]\n",
        "order = 1 #@param [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\"] {type:\"raw\"}\n",
        "chains = \"A\" #@param {type:\"string\"}\n",
        "add_potential = True #@param {type:\"boolean\"}\n",
        "#@markdown - `symmetry='auto'` enables automatic symmetry dectection with [AnAnaS](https://team.inria.fr/nano-d/software/ananas/).\n",
        "#@markdown - `chains=\"A,B\"` filter PDB input to these chains (may help auto-symm detector)\n",
        "#@markdown - `add_potential` to discourage clashes between chains\n",
        "\n",
        "# determine where to save\n",
        "path = name\n",
        "while os.path.exists(f\"outputs/{path}_0.pdb\"):\n",
        "  path = name + \"_\" + ''.join(random.choices(string.ascii_lowercase + string.digits, k=5))\n",
        "\n",
        "flags = {\"contigs\":contigs,\n",
        "         \"pdb\":pdb,\n",
        "         \"order\":order,\n",
        "         \"iterations\":iterations,\n",
        "         \"symmetry\":symmetry,\n",
        "         \"hotspot\":hotspot,\n",
        "         \"path\":path,\n",
        "         \"chains\":chains,\n",
        "         \"add_potential\":add_potential,\n",
        "         \"num_designs\":num_designs,\n",
        "         \"visual\":visual}\n",
        "\n",
        "for k,v in flags.items():\n",
        "  if isinstance(v,str):\n",
        "    flags[k] = v.replace(\"'\",\"\").replace('\"','')\n",
        "\n",
        "contigs, copies = run_diffusion(**flags)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://files.rcsb.org/download/6HN3.pdb -O 6HN3.pdb"
      ],
      "metadata": {
        "id": "vVMe-pjVGsUs",
        "outputId": "8e8a090f-4eec-4c19-9810-e5b5637bbe36",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-05-11 11:30:15--  https://files.rcsb.org/download/6HN3.pdb\n",
            "Resolving files.rcsb.org (files.rcsb.org)... 18.238.217.120, 18.238.217.73, 18.238.217.92, ...\n",
            "Connecting to files.rcsb.org (files.rcsb.org)|18.238.217.120|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/octet-stream]\n",
            "Saving to: ‘6HN3.pdb’\n",
            "\n",
            "6HN3.pdb                [ <=>                ] 307.39K  --.-KB/s    in 0.1s    \n",
            "\n",
            "2025-05-11 11:30:16 (2.74 MB/s) - ‘6HN3.pdb’ saved [314766]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!sed 's/^HETATM/ATOM  /;s/SEC/CYS/g' 6HN3.pdb > 6HN3_cys.pdb"
      ],
      "metadata": {
        "id": "QcEPNuBJHFMO"
      },
      "execution_count": 22,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ipython-input-23-522aa5541724\n",
        "from RFdiffusion import util # Modified import statement to pull util directly from the RFdiffusion module\n",
        "# treat SEC exactly like CYS\n",
        "util.aa2num[\"SEC\"]   = util.aa2num[\"CYS\"]\n",
        "util.aa2long[util.aa2num[\"SEC\"]] = util.aa2long[util.aa2num[\"CYS\"]]"
      ],
      "metadata": {
        "id": "0lsUHh6BHJ9w"
      },
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "FeyoEjCiHJy9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "cellView": "form",
        "id": "wqEi03_qi_g2",
        "outputId": "9b648fa6-54f3-4239-f51f-17f193a0fc2a",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 497
        }
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_17469627220106173\"  style=\"position: relative; width: 640px; height: 480px;\">\n        <p id=\"3dmolwarning_17469627220106173\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_17469627220106173 = null;\nvar warn = document.getElementById(\"3dmolwarning_17469627220106173\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_17469627220106173 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17469627220106173\"),{backgroundColor:\"white\"});\nviewer_17469627220106173.zoomTo();\n\tviewer_17469627220106173.addModel(\"ATOM      1  N   GLY A   1     -22.584  15.080 -23.580  1.00  0.97\\nATOM      2  CA  GLY A   1     -22.594  14.024 -22.574  1.00  0.97\\nATOM      3  C   GLY A   1     -21.336  13.169 -22.660  1.00  0.97\\nATOM      4  O   GLY A   1     -20.812  12.717 -21.642  1.00  0.97\\nATOM      5  N   GLY A   2     -20.750  13.076 -23.799  1.00  0.98\\nATOM      6  CA  GLY A   2     -19.506  12.343 -23.995  1.00  0.98\\nATOM      7  C   GLY A   2     -18.339  13.045 -23.312  1.00  0.98\\nATOM      8  O   GLY A   2     -17.458  12.398 -22.746  1.00  0.98\\nATOM      9  N   GLY A   3     -18.306  14.326 -23.334  1.00  0.99\\nATOM     10  CA  GLY A   3     -17.296  15.103 -22.625  1.00  0.99\\nATOM     11  C   GLY A   3     -17.398  14.897 -21.119  1.00  0.99\\nATOM     12  O   GLY A   3     -16.385  14.809 -20.425  1.00  0.99\\nATOM     13  N   GLY A   4     -18.631  14.720 -20.631  1.00  0.99\\nATOM     14  CA  GLY A   4     -18.857  14.457 -19.215  1.00  0.99\\nATOM     15  C   GLY A   4     -18.320  13.089 -18.815  1.00  0.99\\nATOM     16  O   GLY A   4     -17.746  12.928 -17.738  1.00  0.99\\nATOM     17  N   GLY A   5     -18.494  12.086 -19.705  1.00  0.98\\nATOM     18  CA  GLY A   5     -17.999  10.739 -19.449  1.00  0.98\\nATOM     19  C   GLY A   5     -16.478  10.719 -19.360  1.00  0.98\\nATOM     20  O   GLY A   5     -15.906  10.041 -18.507  1.00  0.98\\nATOM     21  N   GLY A   6     -15.830  11.442 -20.259  1.00  0.98\\nATOM     22  CA  GLY A   6     -14.376  11.542 -20.275  1.00  0.98\\nATOM     23  C   GLY A   6     -13.859  12.259 -19.034  1.00  0.98\\nATOM     24  O   GLY A   6     -12.850  11.863 -18.452  1.00  0.98\\nATOM     25  N   GLY A   7     -14.544  13.354 -18.654  1.00  0.99\\nATOM     26  CA  GLY A   7     -14.177  14.100 -17.456  1.00  0.99\\nATOM     27  C   GLY A   7     -14.353  13.251 -16.203  1.00  0.99\\nATOM     28  O   GLY A   7     -13.526  13.293 -15.292  1.00  0.99\\nATOM     29  N   GLY A   8     -15.429  12.467 -16.149  1.00  0.98\\nATOM     30  CA  GLY A   8     -15.685  11.588 -15.015  1.00  0.98\\nATOM     31  C   GLY A   8     -14.600  10.526 -14.886  1.00  0.98\\nATOM     32  O   GLY A   8     -14.149  10.216 -13.783  1.00  0.98\\nATOM     33  N   GLY A   9     -14.196   9.946 -16.035  1.00  0.98\\nATOM     34  CA  GLY A   9     -13.146   8.934 -16.046  1.00  0.98\\nATOM     35  C   GLY A   9     -11.820   9.512 -15.568  1.00  0.98\\nATOM     36  O   GLY A   9     -11.092   8.875 -14.806  1.00  0.98\\nATOM     37  N   GLY A  10     -11.508  10.722 -16.052  1.00  0.98\\nATOM     38  CA  GLY A  10     -10.280  11.402 -15.658  1.00  0.98\\nATOM     39  C   GLY A  10     -10.288  11.737 -14.172  1.00  0.98\\nATOM     40  O   GLY A  10      -9.275  11.590 -13.488  1.00  0.98\\nATOM     41  N   GLY A  11     -11.445  12.227 -13.672  1.00  0.99\\nATOM     42  CA  GLY A  11     -11.585  12.547 -12.257  1.00  0.99\\nATOM     43  C   GLY A  11     -11.423  11.303 -11.393  1.00  0.99\\nATOM     44  O   GLY A  11     -10.778  11.343 -10.345  1.00  0.99\\nATOM     45  N   GLY A  12     -12.051  10.185 -11.850  1.00  0.98\\nATOM     46  CA  GLY A  12     -11.958   8.921 -11.130  1.00  0.98\\nATOM     47  C   GLY A  12     -10.521   8.417 -11.085  1.00  0.98\\nATOM     48  O   GLY A  12     -10.064   7.904 -10.064  1.00  0.98\\nATOM     49  N   GLY A  13      -9.817   8.575 -12.201  1.00  0.98\\nATOM     50  CA  GLY A  13      -8.416   8.180 -12.276  1.00  0.98\\nATOM     51  C   GLY A  13      -7.561   8.998 -11.316  1.00  0.98\\nATOM     52  O   GLY A  13      -6.674   8.464 -10.651  1.00  0.98\\nATOM     53  N   GLY A  14      -7.843  10.330 -11.284  1.00  0.98\\nATOM     54  CA  GLY A  14      -7.147  11.228 -10.370  1.00  0.98\\nATOM     55  C   GLY A  14      -7.393  10.836  -8.918  1.00  0.98\\nATOM     56  O   GLY A  14      -6.477  10.858  -8.096  1.00  0.98\\nATOM     57  N   GLY A  15      -8.655  10.457  -8.627  1.00  0.98\\nATOM     58  CA  GLY A  15      -9.022  10.021  -7.285  1.00  0.98\\nATOM     59  C   GLY A  15      -8.214   8.801  -6.863  1.00  0.98\\nATOM     60  O   GLY A  15      -7.763   8.708  -5.721  1.00  0.98\\nATOM     61  N   GLY A  16      -8.091   7.819  -7.801  1.00  0.98\\nATOM     62  CA  GLY A  16      -7.335   6.602  -7.529  1.00  0.98\\nATOM     63  C   GLY A  16      -5.866   6.912  -7.268  1.00  0.98\\nATOM     64  O   GLY A  16      -5.254   6.346  -6.363  1.00  0.98\\nATOM     65  N   GLY A  17      -5.322   7.816  -8.038  1.00  0.98\\nATOM     66  CA  GLY A  17      -3.929   8.216  -7.880  1.00  0.98\\nATOM     67  C   GLY A  17      -3.704   8.907  -6.541  1.00  0.98\\nATOM     68  O   GLY A  17      -2.698   8.671  -5.872  1.00  0.98\\nATOM     69  N   GLY A  18      -4.657   9.815  -6.183  1.00  0.98\\nATOM     70  CA  GLY A  18      -4.578  10.526  -4.912  1.00  0.98\\nATOM     71  C   GLY A  18      -4.641   9.560  -3.736  1.00  0.98\\nATOM     72  O   GLY A  18      -3.915   9.714  -2.754  1.00  0.98\\nATOM     73  N   GLY A  19      -5.552   8.534  -3.900  1.00  0.98\\nATOM     74  CA  GLY A  19      -5.695   7.533  -2.850  1.00  0.98\\nATOM     75  C   GLY A  19      -4.417   6.722  -2.684  1.00  0.98\\nATOM     76  O   GLY A  19      -4.000   6.424  -1.565  1.00  0.98\\nATOM     77  N   GLY A  20      -3.806   6.331  -3.811  1.00  0.98\\nATOM     78  CA  GLY A  20      -2.562   5.572  -3.779  1.00  0.98\\nATOM     79  C   GLY A  20      -1.430   6.393  -3.176  1.00  0.98\\nATOM     80  O   GLY A  20      -0.627   5.883  -2.395  1.00  0.98\\nATOM     81  N   GLY A  21      -1.426   7.697  -3.502  1.00  0.98\\nATOM     82  CA  GLY A  21      -0.418   8.601  -2.962  1.00  0.98\\nATOM     83  C   GLY A  21      -0.561   8.748  -1.453  1.00  0.98\\nATOM     84  O   GLY A  21       0.431   8.754  -0.724  1.00  0.98\\nATOM     85  N   GLY A  22      -1.843   8.883  -1.011  1.00  0.98\\nATOM     86  CA  GLY A  22      -2.135   9.029   0.410  1.00  0.98\\nATOM     87  C   GLY A  22      -1.691   7.799   1.193  1.00  0.98\\nATOM     88  O   GLY A  22      -1.151   7.913   2.293  1.00  0.98\\nATOM     89  N   GLY A  23      -1.869   6.618   0.578  1.00  0.97\\nATOM     90  CA  GLY A  23      -1.487   5.367   1.223  1.00  0.97\\nATOM     91  C   GLY A  23       0.028   5.239   1.323  1.00  0.97\\nATOM     92  O   GLY A  23       0.557   4.794   2.341  1.00  0.97\\nATOM     93  N   GLY A  24       0.706   5.558   0.219  1.00  0.97\\nATOM     94  CA  GLY A  24       2.160   5.461   0.178  1.00  0.97\\nATOM     95  C   GLY A  24       2.805   6.433   1.157  1.00  0.97\\nATOM     96  O   GLY A  24       3.778   6.096   1.831  1.00  0.97\\nATOM     97  N   GLY A  25       2.186   7.601   1.295  1.00  0.98\\nATOM     98  CA  GLY A  25       2.673   8.596   2.243  1.00  0.98\\nATOM     99  C   GLY A  25       2.538   8.102   3.677  1.00  0.98\\nATOM    100  O   GLY A  25       3.433   8.299   4.499  1.00  0.98\\nATOM    101  N   GLY A  26       1.394   7.419   3.951  1.00  0.97\\nATOM    102  CA  GLY A  26       1.159   6.857   5.276  1.00  0.97\\nATOM    103  C   GLY A  26       2.133   5.725   5.574  1.00  0.97\\nATOM    104  O   GLY A  26       2.626   5.597   6.695  1.00  0.97\\nATOM    105  N   GLY A  27       2.400   4.919   4.562  1.00  0.97\\nATOM    106  CA  GLY A  27       3.351   3.823   4.708  1.00  0.97\\nATOM    107  C   GLY A  27       4.747   4.344   5.025  1.00  0.97\\nATOM    108  O   GLY A  27       5.450   3.789   5.870  1.00  0.97\\nATOM    109  N   GLY A  28       5.136   5.359   4.270  1.00  0.96\\nATOM    110  CA  GLY A  28       6.444   5.976   4.456  1.00  0.96\\nATOM    111  C   GLY A  28       6.575   6.577   5.850  1.00  0.96\\nATOM    112  O   GLY A  28       7.623   6.467   6.487  1.00  0.96\\nATOM    113  N   GLY A  29       5.514   7.041   6.390  1.00  0.97\\nATOM    114  CA  GLY A  29       5.492   7.647   7.715  1.00  0.97\\nATOM    115  C   GLY A  29       5.268   6.598   8.797  1.00  0.97\\nATOM    116  O   GLY A  29       5.839   6.682   9.884  1.00  0.97\\nATOM    117  N   GLY A  30       4.868   5.360   8.328  1.00  0.96\\nATOM    118  CA  GLY A  30       4.625   4.282   9.279  1.00  0.96\\nATOM    119  C   GLY A  30       3.459   4.613  10.201  1.00  0.96\\nATOM    120  O   GLY A  30       3.197   3.898  11.168  1.00  0.96\\nATOM    121  N   GLY A  31       2.381   5.098   9.640  1.00  0.97\\nATOM    122  CA  GLY A  31       1.212   5.577  10.368  1.00  0.97\\nATOM    123  C   GLY A  31       0.205   4.457  10.591  1.00  0.97\\nATOM    124  O   GLY A  31      -0.483   4.423  11.611  1.00  0.97\\nATOM    125  N   GLY A  32       0.303   3.347  10.002  1.00  0.96\\nATOM    126  CA  GLY A  32      -0.664   2.259  10.093  1.00  0.96\\nATOM    127  C   GLY A  32       0.026   0.932  10.382  1.00  0.96\\nATOM    128  O   GLY A  32       1.198   0.746  10.053  1.00  0.96\\nATOM    129  N   GLY A  33      -0.802  -0.022  10.829  1.00  0.97\\nATOM    130  CA  GLY A  33      -0.274  -1.364  11.045  1.00  0.97\\nATOM    131  C   GLY A  33      -0.161  -2.129   9.733  1.00  0.97\\nATOM    132  O   GLY A  33      -0.901  -1.868   8.785  1.00  0.97\\nATOM    133  N   GLY A  34       0.361  -3.280   9.818  1.00  0.97\\nATOM    134  CA  GLY A  34       0.516  -4.163   8.668  1.00  0.97\\nATOM    135  C   GLY A  34      -0.836  -4.631   8.146  1.00  0.97\\nATOM    136  O   GLY A  34      -1.050  -4.714   6.937  1.00  0.97\\nATOM    137  N   GLY A  35      -1.760  -4.909   9.094  1.00  0.98\\nATOM    138  CA  GLY A  35      -3.091  -5.391   8.743  1.00  0.98\\nATOM    139  C   GLY A  35      -3.860  -4.349   7.942  1.00  0.98\\nATOM    140  O   GLY A  35      -4.536  -4.677   6.967  1.00  0.98\\nATOM    141  N   GLY A  36      -3.761  -3.086   8.359  1.00  0.98\\nATOM    142  CA  GLY A  36      -4.430  -1.976   7.690  1.00  0.98\\nATOM    143  C   GLY A  36      -3.864  -1.751   6.294  1.00  0.98\\nATOM    144  O   GLY A  36      -4.606  -1.496   5.346  1.00  0.98\\nATOM    145  N   GLY A  37      -2.530  -1.858   6.203  1.00  0.97\\nATOM    146  CA  GLY A  37      -1.859  -1.695   4.920  1.00  0.97\\nATOM    147  C   GLY A  37      -2.321  -2.746   3.918  1.00  0.97\\nATOM    148  O   GLY A  37      -2.565  -2.440   2.751  1.00  0.97\\nATOM    149  N   GLY A  38      -2.450  -3.966   4.378  1.00  0.97\\nATOM    150  CA  GLY A  38      -2.869  -5.075   3.529  1.00  0.97\\nATOM    151  C   GLY A  38      -4.305  -4.893   3.054  1.00  0.97\\nATOM    152  O   GLY A  38      -4.622  -5.152   1.893  1.00  0.97\\nATOM    153  N   GLY A  39      -5.196  -4.491   3.954  1.00  0.98\\nATOM    154  CA  GLY A  39      -6.593  -4.239   3.622  1.00  0.98\\nATOM    155  C   GLY A  39      -6.722  -3.142   2.573  1.00  0.98\\nATOM    156  O   GLY A  39      -7.522  -3.249   1.644  1.00  0.98\\nATOM    157  N   GLY A  40      -5.912  -2.092   2.768  1.00  0.98\\nATOM    158  CA  GLY A  40      -5.927  -0.967   1.840  1.00  0.98\\nATOM    159  C   GLY A  40      -5.410  -1.379   0.467  1.00  0.98\\nATOM    160  O   GLY A  40      -5.942  -0.956  -0.559  1.00  0.98\\nATOM    161  N   GLY A  41      -4.445  -2.295   0.463  1.00  0.97\\nATOM    162  CA  GLY A  41      -3.883  -2.785  -0.790  1.00  0.97\\nATOM    163  C   GLY A  41      -4.917  -3.568  -1.589  1.00  0.97\\nATOM    164  O   GLY A  41      -5.015  -3.421  -2.807  1.00  0.97\\nATOM    165  N   GLY A  42      -5.700  -4.377  -0.918  1.00  0.98\\nATOM    166  CA  GLY A  42      -6.735  -5.175  -1.564  1.00  0.98\\nATOM    167  C   GLY A  42      -7.839  -4.292  -2.130  1.00  0.98\\nATOM    168  O   GLY A  42      -8.328  -4.524  -3.236  1.00  0.98\\nATOM    169  N   GLY A  43      -8.209  -3.258  -1.383  1.00  0.98\\nATOM    170  CA  GLY A  43      -9.247  -2.328  -1.811  1.00  0.98\\nATOM    171  C   GLY A  43      -8.798  -1.520  -3.022  1.00  0.98\\nATOM    172  O   GLY A  43      -9.571  -1.298  -3.954  1.00  0.98\\nATOM    173  N   GLY A  44      -7.512  -1.066  -2.975  1.00  0.98\\nATOM    174  CA  GLY A  44      -6.952  -0.291  -4.075  1.00  0.98\\nATOM    175  C   GLY A  44      -6.862  -1.125  -5.347  1.00  0.98\\nATOM    176  O   GLY A  44      -7.142  -0.637  -6.442  1.00  0.98\\nATOM    177  N   GLY A  45      -6.489  -2.378  -5.186  1.00  0.98\\nATOM    178  CA  GLY A  45      -6.390  -3.293  -6.317  1.00  0.98\\nATOM    179  C   GLY A  45      -7.732  -3.444  -7.022  1.00  0.98\\nATOM    180  O   GLY A  45      -7.799  -3.470  -8.251  1.00  0.98\\nATOM    181  N   GLY A  46      -8.817  -3.512  -6.213  1.00  0.98\\nATOM    182  CA  GLY A  46     -10.164  -3.621  -6.761  1.00  0.98\\nATOM    183  C   GLY A  46     -10.527  -2.389  -7.580  1.00  0.98\\nATOM    184  O   GLY A  46     -11.125  -2.498  -8.650  1.00  0.98\\nATOM    185  N   GLY A  47     -10.171  -1.201  -7.022  1.00  0.98\\nATOM    186  CA  GLY A  47     -10.441   0.052  -7.715  1.00  0.98\\nATOM    187  C   GLY A  47      -9.664   0.137  -9.023  1.00  0.98\\nATOM    188  O   GLY A  47     -10.189   0.593 -10.039  1.00  0.98\\nATOM    189  N   GLY A  48      -8.415  -0.343  -8.960  1.00  0.98\\nATOM    190  CA  GLY A  48      -7.571  -0.351 -10.149  1.00  0.98\\nATOM    191  C   GLY A  48      -8.173  -1.218 -11.247  1.00  0.98\\nATOM    192  O   GLY A  48      -8.156  -0.850 -12.421  1.00  0.98\\nATOM    193  N   GLY A  49      -8.697  -2.397 -10.811  1.00  0.98\\nATOM    194  CA  GLY A  49      -9.326  -3.325 -11.743  1.00  0.98\\nATOM    195  C   GLY A  49     -10.526  -2.686 -12.431  1.00  0.98\\nATOM    196  O   GLY A  49     -10.728  -2.858 -13.633  1.00  0.98\\nATOM    197  N   GLY A  50     -11.253  -1.865 -11.695  1.00  0.97\\nATOM    198  CA  GLY A  50     -12.416  -1.169 -12.231  1.00  0.97\\nATOM    199  C   GLY A  50     -12.006  -0.123 -13.260  1.00  0.97\\nATOM    200  O   GLY A  50     -12.641   0.016 -14.305  1.00  0.97\\nATOM    201  N   GLY A  51     -10.944   0.645 -12.916  1.00  0.97\\nATOM    202  CA  GLY A  51     -10.457   1.683 -13.817  1.00  0.97\\nATOM    203  C   GLY A  51      -9.807   1.077 -15.055  1.00  0.97\\nATOM    204  O   GLY A  51      -9.996   1.565 -16.169  1.00  0.97\\nATOM    205  N   GLY A  52      -9.056   0.015 -14.857  1.00  0.97\\nATOM    206  CA  GLY A  52      -8.384  -0.649 -15.967  1.00  0.97\\nATOM    207  C   GLY A  52      -9.391  -1.252 -16.939  1.00  0.97\\nATOM    208  O   GLY A  52      -9.196  -1.212 -18.154  1.00  0.97\\nATOM    209  N   GLY A  53     -10.531  -1.663 -16.390  1.00  0.97\\nATOM    210  CA  GLY A  53     -11.594  -2.240 -17.204  1.00  0.97\\nATOM    211  C   GLY A  53     -12.214  -1.194 -18.123  1.00  0.97\\nATOM    212  O   GLY A  53     -12.542  -1.482 -19.274  1.00  0.97\\nATOM    213  N   GLY A  54     -12.025   0.050 -17.768  1.00  0.97\\nATOM    214  CA  GLY A  54     -12.552   1.157 -18.558  1.00  0.97\\nATOM    215  C   GLY A  54     -11.554   1.598 -19.622  1.00  0.97\\nATOM    216  O   GLY A  54     -11.939   1.985 -20.725  1.00  0.97\\nATOM    217  N   GLY A  55     -10.338   1.309 -19.499  1.00  0.97\\nATOM    218  CA  GLY A  55      -9.283   1.757 -20.400  1.00  0.97\\nATOM    219  C   GLY A  55      -8.961   0.695 -21.443  1.00  0.97\\nATOM    220  O   GLY A  55      -9.854   0.185 -22.120  1.00  0.97\\nATOM    221  N   GLY A  56      -8.028   0.931 -22.256  1.00  0.98\\nATOM    222  CA  GLY A  56      -7.512  -0.074 -23.177  1.00  0.98\\nATOM    223  C   GLY A  56      -6.547  -1.021 -22.476  1.00  0.98\\nATOM    224  O   GLY A  56      -5.955  -0.675 -21.453  1.00  0.98\\nATOM    225  N   GLY A  57      -6.134  -2.028 -23.217  1.00  0.98\\nATOM    226  CA  GLY A  57      -5.226  -3.023 -22.661  1.00  0.98\\nATOM    227  C   GLY A  57      -3.875  -2.406 -22.320  1.00  0.98\\nATOM    228  O   GLY A  57      -3.262  -2.751 -21.309  1.00  0.98\\nATOM    229  N   GLY A  58      -3.431  -1.450 -23.200  1.00  0.99\\nATOM    230  CA  GLY A  58      -2.140  -0.800 -23.005  1.00  0.99\\nATOM    231  C   GLY A  58      -2.113  -0.006 -21.706  1.00  0.99\\nATOM    232  O   GLY A  58      -1.130  -0.042 -20.966  1.00  0.99\\nATOM    233  N   GLY A  59      -3.146   0.779 -21.447  1.00  0.98\\nATOM    234  CA  GLY A  59      -3.238   1.603 -20.247  1.00  0.98\\nATOM    235  C   GLY A  59      -3.502   0.749 -19.013  1.00  0.98\\nATOM    236  O   GLY A  59      -2.967   1.016 -17.937  1.00  0.98\\nATOM    237  N   GLY A  60      -4.250  -0.321 -19.196  1.00  0.98\\nATOM    238  CA  GLY A  60      -4.530  -1.247 -18.105  1.00  0.98\\nATOM    239  C   GLY A  60      -3.254  -1.918 -17.612  1.00  0.98\\nATOM    240  O   GLY A  60      -3.052  -2.080 -16.409  1.00  0.98\\nATOM    241  N   GLY A  61      -2.412  -2.348 -18.562  1.00  0.99\\nATOM    242  CA  GLY A  61      -1.137  -2.970 -18.221  1.00  0.99\\nATOM    243  C   GLY A  61      -0.245  -2.008 -17.447  1.00  0.99\\nATOM    244  O   GLY A  61       0.425  -2.400 -16.492  1.00  0.99\\nATOM    245  N   GLY A  62      -0.289  -0.714 -17.833  1.00  0.98\\nATOM    246  CA  GLY A  62       0.499   0.305 -17.151  1.00  0.98\\nATOM    247  C   GLY A  62       0.007   0.519 -15.726  1.00  0.98\\nATOM    248  O   GLY A  62       0.804   0.660 -14.798  1.00  0.98\\nATOM    249  N   GLY A  63      -1.344   0.644 -15.576  1.00  0.98\\nATOM    250  CA  GLY A  63      -1.939   0.827 -14.257  1.00  0.98\\nATOM    251  C   GLY A  63      -1.654  -0.366 -13.354  1.00  0.98\\nATOM    252  O   GLY A  63      -1.361  -0.203 -12.169  1.00  0.98\\nATOM    253  N   GLY A  64      -1.686  -1.543 -13.926  1.00  0.98\\nATOM    254  CA  GLY A  64      -1.396  -2.766 -13.187  1.00  0.98\\nATOM    255  C   GLY A  64       0.055  -2.800 -12.725  1.00  0.98\\nATOM    256  O   GLY A  64       0.349  -3.214 -11.604  1.00  0.98\\nATOM    257  N   GLY A  65       0.979  -2.340 -13.605  1.00  0.98\\nATOM    258  CA  GLY A  65       2.395  -2.277 -13.263  1.00  0.98\\nATOM    259  C   GLY A  65       2.641  -1.309 -12.113  1.00  0.98\\nATOM    260  O   GLY A  65       3.427  -1.591 -11.209  1.00  0.98\\nATOM    261  N   GLY A  66       1.963  -0.163 -12.148  1.00  0.98\\nATOM    262  CA  GLY A  66       2.067   0.836 -11.091  1.00  0.98\\nATOM    263  C   GLY A  66       1.567   0.285  -9.761  1.00  0.98\\nATOM    264  O   GLY A  66       2.160   0.535  -8.712  1.00  0.98\\nATOM    265  N   GLY A  67       0.460  -0.506  -9.874  1.00  0.98\\nATOM    266  CA  GLY A  67      -0.110  -1.116  -8.679  1.00  0.98\\nATOM    267  C   GLY A  67       0.837  -2.149  -8.082  1.00  0.98\\nATOM    268  O   GLY A  67       0.981  -2.244  -6.863  1.00  0.98\\nATOM    269  N   GLY A  68       1.536  -2.881  -8.953  1.00  0.98\\nATOM    270  CA  GLY A  68       2.511  -3.867  -8.504  1.00  0.98\\nATOM    271  C   GLY A  68       3.609  -3.217  -7.672  1.00  0.98\\nATOM    272  O   GLY A  68       4.029  -3.758  -6.649  1.00  0.98\\nATOM    273  N   GLY A  69       4.050  -2.020  -8.130  1.00  0.98\\nATOM    274  CA  GLY A  69       5.071  -1.271  -7.407  1.00  0.98\\nATOM    275  C   GLY A  69       4.597  -0.907  -6.005  1.00  0.98\\nATOM    276  O   GLY A  69       5.357  -0.996  -5.041  1.00  0.98\\nATOM    277  N   GLY A  70       3.323  -0.449  -5.937  1.00  0.97\\nATOM    278  CA  GLY A  70       2.742  -0.080  -4.652  1.00  0.97\\nATOM    279  C   GLY A  70       2.658  -1.283  -3.721  1.00  0.97\\nATOM    280  O   GLY A  70       2.922  -1.172  -2.523  1.00  0.97\\nATOM    281  N   GLY A  71       2.286  -2.418  -4.293  1.00  0.98\\nATOM    282  CA  GLY A  71       2.182  -3.650  -3.520  1.00  0.98\\nATOM    283  C   GLY A  71       3.533  -4.056  -2.947  1.00  0.98\\nATOM    284  O   GLY A  71       3.628  -4.479  -1.795  1.00  0.98\\nATOM    285  N   GLY A  72       4.583  -3.874  -3.758  1.00  0.98\\nATOM    286  CA  GLY A  72       5.940  -4.186  -3.326  1.00  0.98\\nATOM    287  C   GLY A  72       6.366  -3.299  -2.163  1.00  0.98\\nATOM    288  O   GLY A  72       6.981  -3.767  -1.205  1.00  0.98\\nATOM    289  N   GLY A  73       6.007  -2.005  -2.251  1.00  0.98\\nATOM    290  CA  GLY A  73       6.342  -1.044  -1.207  1.00  0.98\\nATOM    291  C   GLY A  73       5.662  -1.399   0.108  1.00  0.98\\nATOM    292  O   GLY A  73       6.268  -1.309   1.176  1.00  0.98\\nATOM    293  N   GLY A  74       4.383  -1.786  -0.004  1.00  0.97\\nATOM    294  CA  GLY A  74       3.626  -2.176   1.180  1.00  0.97\\nATOM    295  C   GLY A  74       4.238  -3.401   1.846  1.00  0.97\\nATOM    296  O   GLY A  74       4.341  -3.466   3.071  1.00  0.97\\nATOM    297  N   GLY A  75       4.611  -4.380   0.992  1.00  0.98\\nATOM    298  CA  GLY A  75       5.213  -5.605   1.503  1.00  0.98\\nATOM    299  C   GLY A  75       6.546  -5.322   2.184  1.00  0.98\\nATOM    300  O   GLY A  75       6.847  -5.883   3.238  1.00  0.98\\nATOM    301  N   GLY A  76       7.318  -4.415   1.608  1.00  0.98\\nATOM    302  CA  GLY A  76       8.599  -4.019   2.180  1.00  0.98\\nATOM    303  C   GLY A  76       8.415  -3.375   3.549  1.00  0.98\\nATOM    304  O   GLY A  76       9.170  -3.650   4.481  1.00  0.98\\nATOM    305  N   GLY A  77       7.373  -2.490   3.624  1.00  0.98\\nATOM    306  CA  GLY A  77       7.073  -1.805   4.876  1.00  0.98\\nATOM    307  C   GLY A  77       6.606  -2.788   5.943  1.00  0.98\\nATOM    308  O   GLY A  77       6.987  -2.681   7.108  1.00  0.98\\nATOM    309  N   GLY A  78       5.792  -3.752   5.535  1.00  0.98\\nATOM    310  CA  GLY A  78       5.274  -4.759   6.452  1.00  0.98\\nATOM    311  C   GLY A  78       6.392  -5.649   6.982  1.00  0.98\\nATOM    312  O   GLY A  78       6.423  -5.985   8.165  1.00  0.98\\nATOM    313  N   GLY A  79       7.296  -6.057   6.068  1.00  0.98\\nATOM    314  CA  GLY A  79       8.437  -6.874   6.463  1.00  0.98\\nATOM    315  C   GLY A  79       9.328  -6.134   7.453  1.00  0.98\\nATOM    316  O   GLY A  79       9.795  -6.713   8.433  1.00  0.98\\nATOM    317  N   GLY A  80       9.540  -4.846   7.189  1.00  0.98\\nATOM    318  CA  GLY A  80      10.360  -4.012   8.060  1.00  0.98\\nATOM    319  C   GLY A  80       9.733  -3.875   9.441  1.00  0.98\\nATOM    320  O   GLY A  80      10.427  -3.933  10.457  1.00  0.98\\nATOM    321  N   GLY A  81       8.400  -3.648   9.454  1.00  0.98\\nATOM    322  CA  GLY A  81       7.675  -3.506  10.711  1.00  0.98\\nATOM    323  C   GLY A  81       7.732  -4.790  11.528  1.00  0.98\\nATOM    324  O   GLY A  81       7.886  -4.754  12.749  1.00  0.98\\nATOM    325  N   GLY A  82       7.600  -5.914  10.845  1.00  0.98\\nATOM    326  CA  GLY A  82       7.647  -7.210  11.512  1.00  0.98\\nATOM    327  C   GLY A  82       9.019  -7.464  12.124  1.00  0.98\\nATOM    328  O   GLY A  82       9.128  -7.973  13.239  1.00  0.98\\nATOM    329  N   GLY A  83      10.079  -7.183  11.329  1.00  0.99\\nATOM    330  CA  GLY A  83      11.446  -7.364  11.801  1.00  0.99\\nATOM    331  C   GLY A  83      11.735  -6.478  13.006  1.00  0.99\\nATOM    332  O   GLY A  83      12.380  -6.906  13.963  1.00  0.99\\nATOM    333  N   GLY A  84      11.253  -5.244  12.958  1.00  0.99\\nATOM    334  CA  GLY A  84      11.447  -4.298  14.050  1.00  0.99\\nATOM    335  C   GLY A  84      10.790  -4.794  15.332  1.00  0.99\\nATOM    336  O   GLY A  84      11.359  -4.678  16.417  1.00  0.99\\nATOM    337  N   GLY A  85       9.573  -5.302  15.198  1.00  0.99\\nATOM    338  CA  GLY A  85       8.835  -5.822  16.343  1.00  0.99\\nATOM    339  C   GLY A  85       9.517  -7.054  16.924  1.00  0.99\\nATOM    340  O   GLY A  85       9.614  -7.205  18.142  1.00  0.99\\nATOM    341  N   GLY A  86       9.988  -7.927  16.035  1.00  0.99\\nATOM    342  CA  GLY A  86      10.697  -9.130  16.454  1.00  0.99\\nATOM    343  C   GLY A  86      11.995  -8.782  17.170  1.00  0.99\\nATOM    344  O   GLY A  86      12.341  -9.395  18.181  1.00  0.99\\nATOM    345  N   GLY A  87      12.728  -7.794  16.600  1.00  1.00\\nATOM    346  CA  GLY A  87      13.978  -7.333  17.192  1.00  1.00\\nATOM    347  C   GLY A  87      13.751  -6.772  18.590  1.00  1.00\\nATOM    348  O   GLY A  87      14.539  -7.019  19.504  1.00  1.00\\nATOM    349  N   GLY A  88      12.679  -6.016  18.756  1.00  1.00\\nATOM    350  CA  GLY A  88      12.336  -5.425  20.044  1.00  1.00\\nATOM    351  C   GLY A  88      11.962  -6.497  21.059  1.00  1.00\\nATOM    352  O   GLY A  88      12.359  -6.429  22.223  1.00  1.00\\nATOM    353  N   GLY A  89      11.141  -7.452  20.631  1.00  1.00\\nATOM    354  CA  GLY A  89      10.724  -8.553  21.491  1.00  1.00\\nATOM    355  C   GLY A  89      11.918  -9.385  21.939  1.00  1.00\\nATOM    356  O   GLY A  89      12.008  -9.785  23.100  1.00  1.00\\nATOM    357  N   GLY A  90      12.826  -9.666  20.971  1.00  1.00\\nATOM    358  CA  GLY A  90      14.027 -10.438  21.267  1.00  1.00\\nATOM    359  C   GLY A  90      14.921  -9.707  22.259  1.00  1.00\\nATOM    360  O   GLY A  90      15.470 -10.313  23.180  1.00  1.00\\nATOM    361  N   GLY A  91      15.128  -8.403  22.026  1.00  1.00\\nATOM    362  CA  GLY A  91      15.944  -7.583  22.913  1.00  1.00\\nATOM    363  C   GLY A  91      15.377  -7.567  24.326  1.00  1.00\\nATOM    364  O   GLY A  91      16.118  -7.658  25.304  1.00  1.00\\nATOM    365  N   GLY A  92      14.071  -7.421  24.449  1.00  1.00\\nATOM    366  CA  GLY A  92      13.402  -7.425  25.744  1.00  1.00\\nATOM    367  C   GLY A  92      13.571  -8.766  26.447  1.00  1.00\\nATOM    368  O   GLY A  92      13.806  -8.819  27.654  1.00  1.00\\nATOM    369  N   GLY A  93      13.394  -9.860  25.680  1.00  0.99\\nATOM    370  CA  GLY A  93      13.563 -11.202  26.223  1.00  0.99\\nATOM    371  C   GLY A  93      14.978 -11.409  26.749  1.00  0.99\\nATOM    372  O   GLY A  93      15.176 -12.001  27.810  1.00  0.99\\nATOM    373  N   GLY A  94      15.971 -10.939  25.970  1.00  0.99\\nATOM    374  CA  GLY A  94      17.369 -11.049  26.368  1.00  0.99\\nATOM    375  C   GLY A  94      17.628 -10.316  27.678  1.00  0.99\\nATOM    376  O   GLY A  94      18.343 -10.814  28.548  1.00  0.99\\nATOM    377  N   GLY A  95      17.073  -9.106  27.806  1.00  0.99\\nATOM    378  CA  GLY A  95      17.211  -8.320  29.026  1.00  0.99\\nATOM    379  C   GLY A  95      16.581  -9.033  30.215  1.00  0.99\\nATOM    380  O   GLY A  95      17.140  -9.044  31.312  1.00  0.99\\nATOM    381  N   GLY A  96      15.400  -9.616  29.992  1.00  0.99\\nATOM    382  CA  GLY A  96      14.708 -10.362  31.037  1.00  0.99\\nATOM    383  C   GLY A  96      15.518 -11.574  31.477  1.00  0.99\\nATOM    384  O   GLY A  96      15.582 -11.892  32.665  1.00  0.99\\nATOM    385  N   GLY A  97      16.178 -12.234  30.516  1.00  0.99\\nATOM    386  CA  GLY A  97      17.010 -13.393  30.812  1.00  0.99\\nATOM    387  C   GLY A  97      18.213 -13.005  31.663  1.00  0.99\\nATOM    388  O   GLY A  97      18.598 -13.732  32.578  1.00  0.99\\nATOM    389  N   GLY A  98      18.814 -11.845  31.365  1.00  0.99\\nATOM    390  CA  GLY A  98      19.955 -11.344  32.123  1.00  0.99\\nATOM    391  C   GLY A  98      19.563 -11.024  33.560  1.00  0.99\\nATOM    392  O   GLY A  98      20.321 -11.290  34.494  1.00  0.99\\nATOM    393  N   GLY A  99      18.368 -10.597  33.767  1.00  0.99\\nATOM    394  CA  GLY A  99      17.856 -10.280  35.094  1.00  0.99\\nATOM    395  C   GLY A  99      17.478 -11.544  35.856  1.00  0.99\\nATOM    396  O   GLY A  99      17.676 -11.633  37.067  1.00  0.99\\nATOM    397  N   GLY A 100      17.156 -12.582  35.182  1.00  0.98\\nATOM    398  CA  GLY A 100      16.790 -13.853  35.794  1.00  0.98\\nATOM    399  C   GLY A 100      18.022 -14.607  36.277  1.00  0.98\\nATOM    400  O   GLY A 100      18.025 -15.177  37.368  1.00  0.98\",\"pdb\",{\"hbondCutoff\": 4.0});\n\tviewer_17469627220106173.setStyle({\"chain\": \"A\"},{\"cartoon\": {\"color\": \"#33ff33\"}});\n\tviewer_17469627220106173.zoomTo();\nviewer_17469627220106173.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_17469627220106173\"  style=\"position: relative; width: 640px; height: 480px;\">\n",
              "        <p id=\"3dmolwarning_17469627220106173\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    //this is to ignore the existence of requirejs amd\n",
              "    var savedexports, savedmodule;\n",
              "    if (typeof exports !== 'undefined') savedexports = exports;\n",
              "    else exports = {}\n",
              "    if (typeof module !== 'undefined') savedmodule = module;\n",
              "    else module = {}\n",
              "\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "        exports = savedexports;\n",
              "        module = savedmodule;\n",
              "        resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_17469627220106173 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_17469627220106173\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_17469627220106173 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17469627220106173\"),{backgroundColor:\"white\"});\n",
              "viewer_17469627220106173.zoomTo();\n",
              "\tviewer_17469627220106173.addModel(\"ATOM      1  N   GLY A   1     -22.584  15.080 -23.580  1.00  0.97\\nATOM      2  CA  GLY A   1     -22.594  14.024 -22.574  1.00  0.97\\nATOM      3  C   GLY A   1     -21.336  13.169 -22.660  1.00  0.97\\nATOM      4  O   GLY A   1     -20.812  12.717 -21.642  1.00  0.97\\nATOM      5  N   GLY A   2     -20.750  13.076 -23.799  1.00  0.98\\nATOM      6  CA  GLY A   2     -19.506  12.343 -23.995  1.00  0.98\\nATOM      7  C   GLY A   2     -18.339  13.045 -23.312  1.00  0.98\\nATOM      8  O   GLY A   2     -17.458  12.398 -22.746  1.00  0.98\\nATOM      9  N   GLY A   3     -18.306  14.326 -23.334  1.00  0.99\\nATOM     10  CA  GLY A   3     -17.296  15.103 -22.625  1.00  0.99\\nATOM     11  C   GLY A   3     -17.398  14.897 -21.119  1.00  0.99\\nATOM     12  O   GLY A   3     -16.385  14.809 -20.425  1.00  0.99\\nATOM     13  N   GLY A   4     -18.631  14.720 -20.631  1.00  0.99\\nATOM     14  CA  GLY A   4     -18.857  14.457 -19.215  1.00  0.99\\nATOM     15  C   GLY A   4     -18.320  13.089 -18.815  1.00  0.99\\nATOM     16  O   GLY A   4     -17.746  12.928 -17.738  1.00  0.99\\nATOM     17  N   GLY A   5     -18.494  12.086 -19.705  1.00  0.98\\nATOM     18  CA  GLY A   5     -17.999  10.739 -19.449  1.00  0.98\\nATOM     19  C   GLY A   5     -16.478  10.719 -19.360  1.00  0.98\\nATOM     20  O   GLY A   5     -15.906  10.041 -18.507  1.00  0.98\\nATOM     21  N   GLY A   6     -15.830  11.442 -20.259  1.00  0.98\\nATOM     22  CA  GLY A   6     -14.376  11.542 -20.275  1.00  0.98\\nATOM     23  C   GLY A   6     -13.859  12.259 -19.034  1.00  0.98\\nATOM     24  O   GLY A   6     -12.850  11.863 -18.452  1.00  0.98\\nATOM     25  N   GLY A   7     -14.544  13.354 -18.654  1.00  0.99\\nATOM     26  CA  GLY A   7     -14.177  14.100 -17.456  1.00  0.99\\nATOM     27  C   GLY A   7     -14.353  13.251 -16.203  1.00  0.99\\nATOM     28  O   GLY A   7     -13.526  13.293 -15.292  1.00  0.99\\nATOM     29  N   GLY A   8     -15.429  12.467 -16.149  1.00  0.98\\nATOM     30  CA  GLY A   8     -15.685  11.588 -15.015  1.00  0.98\\nATOM     31  C   GLY A   8     -14.600  10.526 -14.886  1.00  0.98\\nATOM     32  O   GLY A   8     -14.149  10.216 -13.783  1.00  0.98\\nATOM     33  N   GLY A   9     -14.196   9.946 -16.035  1.00  0.98\\nATOM     34  CA  GLY A   9     -13.146   8.934 -16.046  1.00  0.98\\nATOM     35  C   GLY A   9     -11.820   9.512 -15.568  1.00  0.98\\nATOM     36  O   GLY A   9     -11.092   8.875 -14.806  1.00  0.98\\nATOM     37  N   GLY A  10     -11.508  10.722 -16.052  1.00  0.98\\nATOM     38  CA  GLY A  10     -10.280  11.402 -15.658  1.00  0.98\\nATOM     39  C   GLY A  10     -10.288  11.737 -14.172  1.00  0.98\\nATOM     40  O   GLY A  10      -9.275  11.590 -13.488  1.00  0.98\\nATOM     41  N   GLY A  11     -11.445  12.227 -13.672  1.00  0.99\\nATOM     42  CA  GLY A  11     -11.585  12.547 -12.257  1.00  0.99\\nATOM     43  C   GLY A  11     -11.423  11.303 -11.393  1.00  0.99\\nATOM     44  O   GLY A  11     -10.778  11.343 -10.345  1.00  0.99\\nATOM     45  N   GLY A  12     -12.051  10.185 -11.850  1.00  0.98\\nATOM     46  CA  GLY A  12     -11.958   8.921 -11.130  1.00  0.98\\nATOM     47  C   GLY A  12     -10.521   8.417 -11.085  1.00  0.98\\nATOM     48  O   GLY A  12     -10.064   7.904 -10.064  1.00  0.98\\nATOM     49  N   GLY A  13      -9.817   8.575 -12.201  1.00  0.98\\nATOM     50  CA  GLY A  13      -8.416   8.180 -12.276  1.00  0.98\\nATOM     51  C   GLY A  13      -7.561   8.998 -11.316  1.00  0.98\\nATOM     52  O   GLY A  13      -6.674   8.464 -10.651  1.00  0.98\\nATOM     53  N   GLY A  14      -7.843  10.330 -11.284  1.00  0.98\\nATOM     54  CA  GLY A  14      -7.147  11.228 -10.370  1.00  0.98\\nATOM     55  C   GLY A  14      -7.393  10.836  -8.918  1.00  0.98\\nATOM     56  O   GLY A  14      -6.477  10.858  -8.096  1.00  0.98\\nATOM     57  N   GLY A  15      -8.655  10.457  -8.627  1.00  0.98\\nATOM     58  CA  GLY A  15      -9.022  10.021  -7.285  1.00  0.98\\nATOM     59  C   GLY A  15      -8.214   8.801  -6.863  1.00  0.98\\nATOM     60  O   GLY A  15      -7.763   8.708  -5.721  1.00  0.98\\nATOM     61  N   GLY A  16      -8.091   7.819  -7.801  1.00  0.98\\nATOM     62  CA  GLY A  16      -7.335   6.602  -7.529  1.00  0.98\\nATOM     63  C   GLY A  16      -5.866   6.912  -7.268  1.00  0.98\\nATOM     64  O   GLY A  16      -5.254   6.346  -6.363  1.00  0.98\\nATOM     65  N   GLY A  17      -5.322   7.816  -8.038  1.00  0.98\\nATOM     66  CA  GLY A  17      -3.929   8.216  -7.880  1.00  0.98\\nATOM     67  C   GLY A  17      -3.704   8.907  -6.541  1.00  0.98\\nATOM     68  O   GLY A  17      -2.698   8.671  -5.872  1.00  0.98\\nATOM     69  N   GLY A  18      -4.657   9.815  -6.183  1.00  0.98\\nATOM     70  CA  GLY A  18      -4.578  10.526  -4.912  1.00  0.98\\nATOM     71  C   GLY A  18      -4.641   9.560  -3.736  1.00  0.98\\nATOM     72  O   GLY A  18      -3.915   9.714  -2.754  1.00  0.98\\nATOM     73  N   GLY A  19      -5.552   8.534  -3.900  1.00  0.98\\nATOM     74  CA  GLY A  19      -5.695   7.533  -2.850  1.00  0.98\\nATOM     75  C   GLY A  19      -4.417   6.722  -2.684  1.00  0.98\\nATOM     76  O   GLY A  19      -4.000   6.424  -1.565  1.00  0.98\\nATOM     77  N   GLY A  20      -3.806   6.331  -3.811  1.00  0.98\\nATOM     78  CA  GLY A  20      -2.562   5.572  -3.779  1.00  0.98\\nATOM     79  C   GLY A  20      -1.430   6.393  -3.176  1.00  0.98\\nATOM     80  O   GLY A  20      -0.627   5.883  -2.395  1.00  0.98\\nATOM     81  N   GLY A  21      -1.426   7.697  -3.502  1.00  0.98\\nATOM     82  CA  GLY A  21      -0.418   8.601  -2.962  1.00  0.98\\nATOM     83  C   GLY A  21      -0.561   8.748  -1.453  1.00  0.98\\nATOM     84  O   GLY A  21       0.431   8.754  -0.724  1.00  0.98\\nATOM     85  N   GLY A  22      -1.843   8.883  -1.011  1.00  0.98\\nATOM     86  CA  GLY A  22      -2.135   9.029   0.410  1.00  0.98\\nATOM     87  C   GLY A  22      -1.691   7.799   1.193  1.00  0.98\\nATOM     88  O   GLY A  22      -1.151   7.913   2.293  1.00  0.98\\nATOM     89  N   GLY A  23      -1.869   6.618   0.578  1.00  0.97\\nATOM     90  CA  GLY A  23      -1.487   5.367   1.223  1.00  0.97\\nATOM     91  C   GLY A  23       0.028   5.239   1.323  1.00  0.97\\nATOM     92  O   GLY A  23       0.557   4.794   2.341  1.00  0.97\\nATOM     93  N   GLY A  24       0.706   5.558   0.219  1.00  0.97\\nATOM     94  CA  GLY A  24       2.160   5.461   0.178  1.00  0.97\\nATOM     95  C   GLY A  24       2.805   6.433   1.157  1.00  0.97\\nATOM     96  O   GLY A  24       3.778   6.096   1.831  1.00  0.97\\nATOM     97  N   GLY A  25       2.186   7.601   1.295  1.00  0.98\\nATOM     98  CA  GLY A  25       2.673   8.596   2.243  1.00  0.98\\nATOM     99  C   GLY A  25       2.538   8.102   3.677  1.00  0.98\\nATOM    100  O   GLY A  25       3.433   8.299   4.499  1.00  0.98\\nATOM    101  N   GLY A  26       1.394   7.419   3.951  1.00  0.97\\nATOM    102  CA  GLY A  26       1.159   6.857   5.276  1.00  0.97\\nATOM    103  C   GLY A  26       2.133   5.725   5.574  1.00  0.97\\nATOM    104  O   GLY A  26       2.626   5.597   6.695  1.00  0.97\\nATOM    105  N   GLY A  27       2.400   4.919   4.562  1.00  0.97\\nATOM    106  CA  GLY A  27       3.351   3.823   4.708  1.00  0.97\\nATOM    107  C   GLY A  27       4.747   4.344   5.025  1.00  0.97\\nATOM    108  O   GLY A  27       5.450   3.789   5.870  1.00  0.97\\nATOM    109  N   GLY A  28       5.136   5.359   4.270  1.00  0.96\\nATOM    110  CA  GLY A  28       6.444   5.976   4.456  1.00  0.96\\nATOM    111  C   GLY A  28       6.575   6.577   5.850  1.00  0.96\\nATOM    112  O   GLY A  28       7.623   6.467   6.487  1.00  0.96\\nATOM    113  N   GLY A  29       5.514   7.041   6.390  1.00  0.97\\nATOM    114  CA  GLY A  29       5.492   7.647   7.715  1.00  0.97\\nATOM    115  C   GLY A  29       5.268   6.598   8.797  1.00  0.97\\nATOM    116  O   GLY A  29       5.839   6.682   9.884  1.00  0.97\\nATOM    117  N   GLY A  30       4.868   5.360   8.328  1.00  0.96\\nATOM    118  CA  GLY A  30       4.625   4.282   9.279  1.00  0.96\\nATOM    119  C   GLY A  30       3.459   4.613  10.201  1.00  0.96\\nATOM    120  O   GLY A  30       3.197   3.898  11.168  1.00  0.96\\nATOM    121  N   GLY A  31       2.381   5.098   9.640  1.00  0.97\\nATOM    122  CA  GLY A  31       1.212   5.577  10.368  1.00  0.97\\nATOM    123  C   GLY A  31       0.205   4.457  10.591  1.00  0.97\\nATOM    124  O   GLY A  31      -0.483   4.423  11.611  1.00  0.97\\nATOM    125  N   GLY A  32       0.303   3.347  10.002  1.00  0.96\\nATOM    126  CA  GLY A  32      -0.664   2.259  10.093  1.00  0.96\\nATOM    127  C   GLY A  32       0.026   0.932  10.382  1.00  0.96\\nATOM    128  O   GLY A  32       1.198   0.746  10.053  1.00  0.96\\nATOM    129  N   GLY A  33      -0.802  -0.022  10.829  1.00  0.97\\nATOM    130  CA  GLY A  33      -0.274  -1.364  11.045  1.00  0.97\\nATOM    131  C   GLY A  33      -0.161  -2.129   9.733  1.00  0.97\\nATOM    132  O   GLY A  33      -0.901  -1.868   8.785  1.00  0.97\\nATOM    133  N   GLY A  34       0.361  -3.280   9.818  1.00  0.97\\nATOM    134  CA  GLY A  34       0.516  -4.163   8.668  1.00  0.97\\nATOM    135  C   GLY A  34      -0.836  -4.631   8.146  1.00  0.97\\nATOM    136  O   GLY A  34      -1.050  -4.714   6.937  1.00  0.97\\nATOM    137  N   GLY A  35      -1.760  -4.909   9.094  1.00  0.98\\nATOM    138  CA  GLY A  35      -3.091  -5.391   8.743  1.00  0.98\\nATOM    139  C   GLY A  35      -3.860  -4.349   7.942  1.00  0.98\\nATOM    140  O   GLY A  35      -4.536  -4.677   6.967  1.00  0.98\\nATOM    141  N   GLY A  36      -3.761  -3.086   8.359  1.00  0.98\\nATOM    142  CA  GLY A  36      -4.430  -1.976   7.690  1.00  0.98\\nATOM    143  C   GLY A  36      -3.864  -1.751   6.294  1.00  0.98\\nATOM    144  O   GLY A  36      -4.606  -1.496   5.346  1.00  0.98\\nATOM    145  N   GLY A  37      -2.530  -1.858   6.203  1.00  0.97\\nATOM    146  CA  GLY A  37      -1.859  -1.695   4.920  1.00  0.97\\nATOM    147  C   GLY A  37      -2.321  -2.746   3.918  1.00  0.97\\nATOM    148  O   GLY A  37      -2.565  -2.440   2.751  1.00  0.97\\nATOM    149  N   GLY A  38      -2.450  -3.966   4.378  1.00  0.97\\nATOM    150  CA  GLY A  38      -2.869  -5.075   3.529  1.00  0.97\\nATOM    151  C   GLY A  38      -4.305  -4.893   3.054  1.00  0.97\\nATOM    152  O   GLY A  38      -4.622  -5.152   1.893  1.00  0.97\\nATOM    153  N   GLY A  39      -5.196  -4.491   3.954  1.00  0.98\\nATOM    154  CA  GLY A  39      -6.593  -4.239   3.622  1.00  0.98\\nATOM    155  C   GLY A  39      -6.722  -3.142   2.573  1.00  0.98\\nATOM    156  O   GLY A  39      -7.522  -3.249   1.644  1.00  0.98\\nATOM    157  N   GLY A  40      -5.912  -2.092   2.768  1.00  0.98\\nATOM    158  CA  GLY A  40      -5.927  -0.967   1.840  1.00  0.98\\nATOM    159  C   GLY A  40      -5.410  -1.379   0.467  1.00  0.98\\nATOM    160  O   GLY A  40      -5.942  -0.956  -0.559  1.00  0.98\\nATOM    161  N   GLY A  41      -4.445  -2.295   0.463  1.00  0.97\\nATOM    162  CA  GLY A  41      -3.883  -2.785  -0.790  1.00  0.97\\nATOM    163  C   GLY A  41      -4.917  -3.568  -1.589  1.00  0.97\\nATOM    164  O   GLY A  41      -5.015  -3.421  -2.807  1.00  0.97\\nATOM    165  N   GLY A  42      -5.700  -4.377  -0.918  1.00  0.98\\nATOM    166  CA  GLY A  42      -6.735  -5.175  -1.564  1.00  0.98\\nATOM    167  C   GLY A  42      -7.839  -4.292  -2.130  1.00  0.98\\nATOM    168  O   GLY A  42      -8.328  -4.524  -3.236  1.00  0.98\\nATOM    169  N   GLY A  43      -8.209  -3.258  -1.383  1.00  0.98\\nATOM    170  CA  GLY A  43      -9.247  -2.328  -1.811  1.00  0.98\\nATOM    171  C   GLY A  43      -8.798  -1.520  -3.022  1.00  0.98\\nATOM    172  O   GLY A  43      -9.571  -1.298  -3.954  1.00  0.98\\nATOM    173  N   GLY A  44      -7.512  -1.066  -2.975  1.00  0.98\\nATOM    174  CA  GLY A  44      -6.952  -0.291  -4.075  1.00  0.98\\nATOM    175  C   GLY A  44      -6.862  -1.125  -5.347  1.00  0.98\\nATOM    176  O   GLY A  44      -7.142  -0.637  -6.442  1.00  0.98\\nATOM    177  N   GLY A  45      -6.489  -2.378  -5.186  1.00  0.98\\nATOM    178  CA  GLY A  45      -6.390  -3.293  -6.317  1.00  0.98\\nATOM    179  C   GLY A  45      -7.732  -3.444  -7.022  1.00  0.98\\nATOM    180  O   GLY A  45      -7.799  -3.470  -8.251  1.00  0.98\\nATOM    181  N   GLY A  46      -8.817  -3.512  -6.213  1.00  0.98\\nATOM    182  CA  GLY A  46     -10.164  -3.621  -6.761  1.00  0.98\\nATOM    183  C   GLY A  46     -10.527  -2.389  -7.580  1.00  0.98\\nATOM    184  O   GLY A  46     -11.125  -2.498  -8.650  1.00  0.98\\nATOM    185  N   GLY A  47     -10.171  -1.201  -7.022  1.00  0.98\\nATOM    186  CA  GLY A  47     -10.441   0.052  -7.715  1.00  0.98\\nATOM    187  C   GLY A  47      -9.664   0.137  -9.023  1.00  0.98\\nATOM    188  O   GLY A  47     -10.189   0.593 -10.039  1.00  0.98\\nATOM    189  N   GLY A  48      -8.415  -0.343  -8.960  1.00  0.98\\nATOM    190  CA  GLY A  48      -7.571  -0.351 -10.149  1.00  0.98\\nATOM    191  C   GLY A  48      -8.173  -1.218 -11.247  1.00  0.98\\nATOM    192  O   GLY A  48      -8.156  -0.850 -12.421  1.00  0.98\\nATOM    193  N   GLY A  49      -8.697  -2.397 -10.811  1.00  0.98\\nATOM    194  CA  GLY A  49      -9.326  -3.325 -11.743  1.00  0.98\\nATOM    195  C   GLY A  49     -10.526  -2.686 -12.431  1.00  0.98\\nATOM    196  O   GLY A  49     -10.728  -2.858 -13.633  1.00  0.98\\nATOM    197  N   GLY A  50     -11.253  -1.865 -11.695  1.00  0.97\\nATOM    198  CA  GLY A  50     -12.416  -1.169 -12.231  1.00  0.97\\nATOM    199  C   GLY A  50     -12.006  -0.123 -13.260  1.00  0.97\\nATOM    200  O   GLY A  50     -12.641   0.016 -14.305  1.00  0.97\\nATOM    201  N   GLY A  51     -10.944   0.645 -12.916  1.00  0.97\\nATOM    202  CA  GLY A  51     -10.457   1.683 -13.817  1.00  0.97\\nATOM    203  C   GLY A  51      -9.807   1.077 -15.055  1.00  0.97\\nATOM    204  O   GLY A  51      -9.996   1.565 -16.169  1.00  0.97\\nATOM    205  N   GLY A  52      -9.056   0.015 -14.857  1.00  0.97\\nATOM    206  CA  GLY A  52      -8.384  -0.649 -15.967  1.00  0.97\\nATOM    207  C   GLY A  52      -9.391  -1.252 -16.939  1.00  0.97\\nATOM    208  O   GLY A  52      -9.196  -1.212 -18.154  1.00  0.97\\nATOM    209  N   GLY A  53     -10.531  -1.663 -16.390  1.00  0.97\\nATOM    210  CA  GLY A  53     -11.594  -2.240 -17.204  1.00  0.97\\nATOM    211  C   GLY A  53     -12.214  -1.194 -18.123  1.00  0.97\\nATOM    212  O   GLY A  53     -12.542  -1.482 -19.274  1.00  0.97\\nATOM    213  N   GLY A  54     -12.025   0.050 -17.768  1.00  0.97\\nATOM    214  CA  GLY A  54     -12.552   1.157 -18.558  1.00  0.97\\nATOM    215  C   GLY A  54     -11.554   1.598 -19.622  1.00  0.97\\nATOM    216  O   GLY A  54     -11.939   1.985 -20.725  1.00  0.97\\nATOM    217  N   GLY A  55     -10.338   1.309 -19.499  1.00  0.97\\nATOM    218  CA  GLY A  55      -9.283   1.757 -20.400  1.00  0.97\\nATOM    219  C   GLY A  55      -8.961   0.695 -21.443  1.00  0.97\\nATOM    220  O   GLY A  55      -9.854   0.185 -22.120  1.00  0.97\\nATOM    221  N   GLY A  56      -8.028   0.931 -22.256  1.00  0.98\\nATOM    222  CA  GLY A  56      -7.512  -0.074 -23.177  1.00  0.98\\nATOM    223  C   GLY A  56      -6.547  -1.021 -22.476  1.00  0.98\\nATOM    224  O   GLY A  56      -5.955  -0.675 -21.453  1.00  0.98\\nATOM    225  N   GLY A  57      -6.134  -2.028 -23.217  1.00  0.98\\nATOM    226  CA  GLY A  57      -5.226  -3.023 -22.661  1.00  0.98\\nATOM    227  C   GLY A  57      -3.875  -2.406 -22.320  1.00  0.98\\nATOM    228  O   GLY A  57      -3.262  -2.751 -21.309  1.00  0.98\\nATOM    229  N   GLY A  58      -3.431  -1.450 -23.200  1.00  0.99\\nATOM    230  CA  GLY A  58      -2.140  -0.800 -23.005  1.00  0.99\\nATOM    231  C   GLY A  58      -2.113  -0.006 -21.706  1.00  0.99\\nATOM    232  O   GLY A  58      -1.130  -0.042 -20.966  1.00  0.99\\nATOM    233  N   GLY A  59      -3.146   0.779 -21.447  1.00  0.98\\nATOM    234  CA  GLY A  59      -3.238   1.603 -20.247  1.00  0.98\\nATOM    235  C   GLY A  59      -3.502   0.749 -19.013  1.00  0.98\\nATOM    236  O   GLY A  59      -2.967   1.016 -17.937  1.00  0.98\\nATOM    237  N   GLY A  60      -4.250  -0.321 -19.196  1.00  0.98\\nATOM    238  CA  GLY A  60      -4.530  -1.247 -18.105  1.00  0.98\\nATOM    239  C   GLY A  60      -3.254  -1.918 -17.612  1.00  0.98\\nATOM    240  O   GLY A  60      -3.052  -2.080 -16.409  1.00  0.98\\nATOM    241  N   GLY A  61      -2.412  -2.348 -18.562  1.00  0.99\\nATOM    242  CA  GLY A  61      -1.137  -2.970 -18.221  1.00  0.99\\nATOM    243  C   GLY A  61      -0.245  -2.008 -17.447  1.00  0.99\\nATOM    244  O   GLY A  61       0.425  -2.400 -16.492  1.00  0.99\\nATOM    245  N   GLY A  62      -0.289  -0.714 -17.833  1.00  0.98\\nATOM    246  CA  GLY A  62       0.499   0.305 -17.151  1.00  0.98\\nATOM    247  C   GLY A  62       0.007   0.519 -15.726  1.00  0.98\\nATOM    248  O   GLY A  62       0.804   0.660 -14.798  1.00  0.98\\nATOM    249  N   GLY A  63      -1.344   0.644 -15.576  1.00  0.98\\nATOM    250  CA  GLY A  63      -1.939   0.827 -14.257  1.00  0.98\\nATOM    251  C   GLY A  63      -1.654  -0.366 -13.354  1.00  0.98\\nATOM    252  O   GLY A  63      -1.361  -0.203 -12.169  1.00  0.98\\nATOM    253  N   GLY A  64      -1.686  -1.543 -13.926  1.00  0.98\\nATOM    254  CA  GLY A  64      -1.396  -2.766 -13.187  1.00  0.98\\nATOM    255  C   GLY A  64       0.055  -2.800 -12.725  1.00  0.98\\nATOM    256  O   GLY A  64       0.349  -3.214 -11.604  1.00  0.98\\nATOM    257  N   GLY A  65       0.979  -2.340 -13.605  1.00  0.98\\nATOM    258  CA  GLY A  65       2.395  -2.277 -13.263  1.00  0.98\\nATOM    259  C   GLY A  65       2.641  -1.309 -12.113  1.00  0.98\\nATOM    260  O   GLY A  65       3.427  -1.591 -11.209  1.00  0.98\\nATOM    261  N   GLY A  66       1.963  -0.163 -12.148  1.00  0.98\\nATOM    262  CA  GLY A  66       2.067   0.836 -11.091  1.00  0.98\\nATOM    263  C   GLY A  66       1.567   0.285  -9.761  1.00  0.98\\nATOM    264  O   GLY A  66       2.160   0.535  -8.712  1.00  0.98\\nATOM    265  N   GLY A  67       0.460  -0.506  -9.874  1.00  0.98\\nATOM    266  CA  GLY A  67      -0.110  -1.116  -8.679  1.00  0.98\\nATOM    267  C   GLY A  67       0.837  -2.149  -8.082  1.00  0.98\\nATOM    268  O   GLY A  67       0.981  -2.244  -6.863  1.00  0.98\\nATOM    269  N   GLY A  68       1.536  -2.881  -8.953  1.00  0.98\\nATOM    270  CA  GLY A  68       2.511  -3.867  -8.504  1.00  0.98\\nATOM    271  C   GLY A  68       3.609  -3.217  -7.672  1.00  0.98\\nATOM    272  O   GLY A  68       4.029  -3.758  -6.649  1.00  0.98\\nATOM    273  N   GLY A  69       4.050  -2.020  -8.130  1.00  0.98\\nATOM    274  CA  GLY A  69       5.071  -1.271  -7.407  1.00  0.98\\nATOM    275  C   GLY A  69       4.597  -0.907  -6.005  1.00  0.98\\nATOM    276  O   GLY A  69       5.357  -0.996  -5.041  1.00  0.98\\nATOM    277  N   GLY A  70       3.323  -0.449  -5.937  1.00  0.97\\nATOM    278  CA  GLY A  70       2.742  -0.080  -4.652  1.00  0.97\\nATOM    279  C   GLY A  70       2.658  -1.283  -3.721  1.00  0.97\\nATOM    280  O   GLY A  70       2.922  -1.172  -2.523  1.00  0.97\\nATOM    281  N   GLY A  71       2.286  -2.418  -4.293  1.00  0.98\\nATOM    282  CA  GLY A  71       2.182  -3.650  -3.520  1.00  0.98\\nATOM    283  C   GLY A  71       3.533  -4.056  -2.947  1.00  0.98\\nATOM    284  O   GLY A  71       3.628  -4.479  -1.795  1.00  0.98\\nATOM    285  N   GLY A  72       4.583  -3.874  -3.758  1.00  0.98\\nATOM    286  CA  GLY A  72       5.940  -4.186  -3.326  1.00  0.98\\nATOM    287  C   GLY A  72       6.366  -3.299  -2.163  1.00  0.98\\nATOM    288  O   GLY A  72       6.981  -3.767  -1.205  1.00  0.98\\nATOM    289  N   GLY A  73       6.007  -2.005  -2.251  1.00  0.98\\nATOM    290  CA  GLY A  73       6.342  -1.044  -1.207  1.00  0.98\\nATOM    291  C   GLY A  73       5.662  -1.399   0.108  1.00  0.98\\nATOM    292  O   GLY A  73       6.268  -1.309   1.176  1.00  0.98\\nATOM    293  N   GLY A  74       4.383  -1.786  -0.004  1.00  0.97\\nATOM    294  CA  GLY A  74       3.626  -2.176   1.180  1.00  0.97\\nATOM    295  C   GLY A  74       4.238  -3.401   1.846  1.00  0.97\\nATOM    296  O   GLY A  74       4.341  -3.466   3.071  1.00  0.97\\nATOM    297  N   GLY A  75       4.611  -4.380   0.992  1.00  0.98\\nATOM    298  CA  GLY A  75       5.213  -5.605   1.503  1.00  0.98\\nATOM    299  C   GLY A  75       6.546  -5.322   2.184  1.00  0.98\\nATOM    300  O   GLY A  75       6.847  -5.883   3.238  1.00  0.98\\nATOM    301  N   GLY A  76       7.318  -4.415   1.608  1.00  0.98\\nATOM    302  CA  GLY A  76       8.599  -4.019   2.180  1.00  0.98\\nATOM    303  C   GLY A  76       8.415  -3.375   3.549  1.00  0.98\\nATOM    304  O   GLY A  76       9.170  -3.650   4.481  1.00  0.98\\nATOM    305  N   GLY A  77       7.373  -2.490   3.624  1.00  0.98\\nATOM    306  CA  GLY A  77       7.073  -1.805   4.876  1.00  0.98\\nATOM    307  C   GLY A  77       6.606  -2.788   5.943  1.00  0.98\\nATOM    308  O   GLY A  77       6.987  -2.681   7.108  1.00  0.98\\nATOM    309  N   GLY A  78       5.792  -3.752   5.535  1.00  0.98\\nATOM    310  CA  GLY A  78       5.274  -4.759   6.452  1.00  0.98\\nATOM    311  C   GLY A  78       6.392  -5.649   6.982  1.00  0.98\\nATOM    312  O   GLY A  78       6.423  -5.985   8.165  1.00  0.98\\nATOM    313  N   GLY A  79       7.296  -6.057   6.068  1.00  0.98\\nATOM    314  CA  GLY A  79       8.437  -6.874   6.463  1.00  0.98\\nATOM    315  C   GLY A  79       9.328  -6.134   7.453  1.00  0.98\\nATOM    316  O   GLY A  79       9.795  -6.713   8.433  1.00  0.98\\nATOM    317  N   GLY A  80       9.540  -4.846   7.189  1.00  0.98\\nATOM    318  CA  GLY A  80      10.360  -4.012   8.060  1.00  0.98\\nATOM    319  C   GLY A  80       9.733  -3.875   9.441  1.00  0.98\\nATOM    320  O   GLY A  80      10.427  -3.933  10.457  1.00  0.98\\nATOM    321  N   GLY A  81       8.400  -3.648   9.454  1.00  0.98\\nATOM    322  CA  GLY A  81       7.675  -3.506  10.711  1.00  0.98\\nATOM    323  C   GLY A  81       7.732  -4.790  11.528  1.00  0.98\\nATOM    324  O   GLY A  81       7.886  -4.754  12.749  1.00  0.98\\nATOM    325  N   GLY A  82       7.600  -5.914  10.845  1.00  0.98\\nATOM    326  CA  GLY A  82       7.647  -7.210  11.512  1.00  0.98\\nATOM    327  C   GLY A  82       9.019  -7.464  12.124  1.00  0.98\\nATOM    328  O   GLY A  82       9.128  -7.973  13.239  1.00  0.98\\nATOM    329  N   GLY A  83      10.079  -7.183  11.329  1.00  0.99\\nATOM    330  CA  GLY A  83      11.446  -7.364  11.801  1.00  0.99\\nATOM    331  C   GLY A  83      11.735  -6.478  13.006  1.00  0.99\\nATOM    332  O   GLY A  83      12.380  -6.906  13.963  1.00  0.99\\nATOM    333  N   GLY A  84      11.253  -5.244  12.958  1.00  0.99\\nATOM    334  CA  GLY A  84      11.447  -4.298  14.050  1.00  0.99\\nATOM    335  C   GLY A  84      10.790  -4.794  15.332  1.00  0.99\\nATOM    336  O   GLY A  84      11.359  -4.678  16.417  1.00  0.99\\nATOM    337  N   GLY A  85       9.573  -5.302  15.198  1.00  0.99\\nATOM    338  CA  GLY A  85       8.835  -5.822  16.343  1.00  0.99\\nATOM    339  C   GLY A  85       9.517  -7.054  16.924  1.00  0.99\\nATOM    340  O   GLY A  85       9.614  -7.205  18.142  1.00  0.99\\nATOM    341  N   GLY A  86       9.988  -7.927  16.035  1.00  0.99\\nATOM    342  CA  GLY A  86      10.697  -9.130  16.454  1.00  0.99\\nATOM    343  C   GLY A  86      11.995  -8.782  17.170  1.00  0.99\\nATOM    344  O   GLY A  86      12.341  -9.395  18.181  1.00  0.99\\nATOM    345  N   GLY A  87      12.728  -7.794  16.600  1.00  1.00\\nATOM    346  CA  GLY A  87      13.978  -7.333  17.192  1.00  1.00\\nATOM    347  C   GLY A  87      13.751  -6.772  18.590  1.00  1.00\\nATOM    348  O   GLY A  87      14.539  -7.019  19.504  1.00  1.00\\nATOM    349  N   GLY A  88      12.679  -6.016  18.756  1.00  1.00\\nATOM    350  CA  GLY A  88      12.336  -5.425  20.044  1.00  1.00\\nATOM    351  C   GLY A  88      11.962  -6.497  21.059  1.00  1.00\\nATOM    352  O   GLY A  88      12.359  -6.429  22.223  1.00  1.00\\nATOM    353  N   GLY A  89      11.141  -7.452  20.631  1.00  1.00\\nATOM    354  CA  GLY A  89      10.724  -8.553  21.491  1.00  1.00\\nATOM    355  C   GLY A  89      11.918  -9.385  21.939  1.00  1.00\\nATOM    356  O   GLY A  89      12.008  -9.785  23.100  1.00  1.00\\nATOM    357  N   GLY A  90      12.826  -9.666  20.971  1.00  1.00\\nATOM    358  CA  GLY A  90      14.027 -10.438  21.267  1.00  1.00\\nATOM    359  C   GLY A  90      14.921  -9.707  22.259  1.00  1.00\\nATOM    360  O   GLY A  90      15.470 -10.313  23.180  1.00  1.00\\nATOM    361  N   GLY A  91      15.128  -8.403  22.026  1.00  1.00\\nATOM    362  CA  GLY A  91      15.944  -7.583  22.913  1.00  1.00\\nATOM    363  C   GLY A  91      15.377  -7.567  24.326  1.00  1.00\\nATOM    364  O   GLY A  91      16.118  -7.658  25.304  1.00  1.00\\nATOM    365  N   GLY A  92      14.071  -7.421  24.449  1.00  1.00\\nATOM    366  CA  GLY A  92      13.402  -7.425  25.744  1.00  1.00\\nATOM    367  C   GLY A  92      13.571  -8.766  26.447  1.00  1.00\\nATOM    368  O   GLY A  92      13.806  -8.819  27.654  1.00  1.00\\nATOM    369  N   GLY A  93      13.394  -9.860  25.680  1.00  0.99\\nATOM    370  CA  GLY A  93      13.563 -11.202  26.223  1.00  0.99\\nATOM    371  C   GLY A  93      14.978 -11.409  26.749  1.00  0.99\\nATOM    372  O   GLY A  93      15.176 -12.001  27.810  1.00  0.99\\nATOM    373  N   GLY A  94      15.971 -10.939  25.970  1.00  0.99\\nATOM    374  CA  GLY A  94      17.369 -11.049  26.368  1.00  0.99\\nATOM    375  C   GLY A  94      17.628 -10.316  27.678  1.00  0.99\\nATOM    376  O   GLY A  94      18.343 -10.814  28.548  1.00  0.99\\nATOM    377  N   GLY A  95      17.073  -9.106  27.806  1.00  0.99\\nATOM    378  CA  GLY A  95      17.211  -8.320  29.026  1.00  0.99\\nATOM    379  C   GLY A  95      16.581  -9.033  30.215  1.00  0.99\\nATOM    380  O   GLY A  95      17.140  -9.044  31.312  1.00  0.99\\nATOM    381  N   GLY A  96      15.400  -9.616  29.992  1.00  0.99\\nATOM    382  CA  GLY A  96      14.708 -10.362  31.037  1.00  0.99\\nATOM    383  C   GLY A  96      15.518 -11.574  31.477  1.00  0.99\\nATOM    384  O   GLY A  96      15.582 -11.892  32.665  1.00  0.99\\nATOM    385  N   GLY A  97      16.178 -12.234  30.516  1.00  0.99\\nATOM    386  CA  GLY A  97      17.010 -13.393  30.812  1.00  0.99\\nATOM    387  C   GLY A  97      18.213 -13.005  31.663  1.00  0.99\\nATOM    388  O   GLY A  97      18.598 -13.732  32.578  1.00  0.99\\nATOM    389  N   GLY A  98      18.814 -11.845  31.365  1.00  0.99\\nATOM    390  CA  GLY A  98      19.955 -11.344  32.123  1.00  0.99\\nATOM    391  C   GLY A  98      19.563 -11.024  33.560  1.00  0.99\\nATOM    392  O   GLY A  98      20.321 -11.290  34.494  1.00  0.99\\nATOM    393  N   GLY A  99      18.368 -10.597  33.767  1.00  0.99\\nATOM    394  CA  GLY A  99      17.856 -10.280  35.094  1.00  0.99\\nATOM    395  C   GLY A  99      17.478 -11.544  35.856  1.00  0.99\\nATOM    396  O   GLY A  99      17.676 -11.633  37.067  1.00  0.99\\nATOM    397  N   GLY A 100      17.156 -12.582  35.182  1.00  0.98\\nATOM    398  CA  GLY A 100      16.790 -13.853  35.794  1.00  0.98\\nATOM    399  C   GLY A 100      18.022 -14.607  36.277  1.00  0.98\\nATOM    400  O   GLY A 100      18.025 -15.177  37.368  1.00  0.98\",\"pdb\",{\"hbondCutoff\": 4.0});\n",
              "\tviewer_17469627220106173.setStyle({\"chain\": \"A\"},{\"cartoon\": {\"color\": \"#33ff33\"}});\n",
              "\tviewer_17469627220106173.zoomTo();\n",
              "viewer_17469627220106173.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "#@title Display 3D structure {run: \"auto\"}\n",
        "animate = \"none\" #@param [\"none\", \"movie\", \"interactive\"]\n",
        "color = \"chain\" #@param [\"rainbow\", \"chain\", \"plddt\"]\n",
        "denoise = True\n",
        "dpi = 100 #@param [\"100\", \"200\", \"400\"] {type:\"raw\"}\n",
        "from colabdesign.shared.plot import pymol_color_list\n",
        "from colabdesign.rf.utils import get_ca, get_Ls, make_animation\n",
        "from string import ascii_uppercase,ascii_lowercase\n",
        "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
        "\n",
        "def plot_pdb(num=0):\n",
        "  if denoise:\n",
        "    pdb_traj = f\"outputs/traj/{path}_{num}_pX0_traj.pdb\"\n",
        "  else:\n",
        "    pdb_traj = f\"outputs/traj/{path}_{num}_Xt-1_traj.pdb\"\n",
        "  if animate in [\"none\",\"interactive\"]:\n",
        "    hbondCutoff = 4.0\n",
        "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "    if animate == \"interactive\":\n",
        "      pdb_str = open(pdb_traj,'r').read()\n",
        "      view.addModelsAsFrames(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "    else:\n",
        "      pdb = f\"outputs/{path}_{num}.pdb\"\n",
        "      pdb_str = open(pdb,'r').read()\n",
        "      view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "    if color == \"rainbow\":\n",
        "      view.setStyle({'cartoon': {'color':'spectrum'}})\n",
        "    elif color == \"chain\":\n",
        "      for n,chain,c in zip(range(len(contigs)),\n",
        "                              alphabet_list,\n",
        "                              pymol_color_list):\n",
        "          view.setStyle({'chain':chain},{'cartoon': {'color':c}})\n",
        "    else:\n",
        "      view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':0.5,'max':0.9}}})\n",
        "    view.zoomTo()\n",
        "    if animate == \"interactive\":\n",
        "      view.animate({'loop': 'backAndForth'})\n",
        "    view.show()\n",
        "  else:\n",
        "    Ls = get_Ls(contigs)\n",
        "    xyz, bfact = get_ca(pdb_traj, get_bfact=True)\n",
        "    xyz = xyz.reshape((-1,sum(Ls),3))[::-1]\n",
        "    bfact = bfact.reshape((-1,sum(Ls)))[::-1]\n",
        "    if color == \"chain\":\n",
        "      display(HTML(make_animation(xyz, Ls=Ls, dpi=dpi, ref=-1)))\n",
        "    elif color == \"rainbow\":\n",
        "      display(HTML(make_animation(xyz, dpi=dpi, ref=-1)))\n",
        "    else:\n",
        "      display(HTML(make_animation(xyz, plddt=bfact*100, dpi=dpi, ref=-1)))\n",
        "\n",
        "\n",
        "if num_designs > 1:\n",
        "  output = widgets.Output()\n",
        "  def on_change(change):\n",
        "    if change['name'] == 'value':\n",
        "      with output:\n",
        "        output.clear_output(wait=True)\n",
        "        plot_pdb(change['new'])\n",
        "  dropdown = widgets.Dropdown(\n",
        "      options=[(f'{k}',k) for k in range(num_designs)],\n",
        "      value=0, description='design:',\n",
        "  )\n",
        "  dropdown.observe(on_change)\n",
        "  display(widgets.VBox([dropdown, output]))\n",
        "  with output:\n",
        "    plot_pdb(dropdown.value)\n",
        "else:\n",
        "  plot_pdb()"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "alVF0UnyHI4l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "rES3p-q6j4tc"
      },
      "outputs": [],
      "source": [
        "%%time\n",
        "#@title run **ProteinMPNN** to generate a sequence and **AlphaFold** to validate\n",
        "num_seqs = 8 #@param [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\", \"64\"] {type:\"raw\"}\n",
        "initial_guess = False #@param {type:\"boolean\"}\n",
        "num_recycles = 1 #@param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\"] {type:\"raw\"}\n",
        "use_multimer = False #@param {type:\"boolean\"}\n",
        "rm_aa = \"C\" #@param {type:\"string\"}\n",
        "mpnn_sampling_temp = 0.1 #@param [\"0.0001\", \"0.1\", \"0.15\", \"0.2\", \"0.25\", \"0.3\", \"0.5\", \"1.0\"] {type:\"raw\"}\n",
        "#@markdown - for **binder** design, we recommend `initial_guess=True num_recycles=3`\n",
        "\n",
        "if not os.path.isfile(\"params/done.txt\"):\n",
        "  print(\"downloading AlphaFold params...\")\n",
        "  while not os.path.isfile(\"params/done.txt\"):\n",
        "    time.sleep(5)\n",
        "\n",
        "contigs_str = \":\".join(contigs)\n",
        "opts = [f\"--pdb=outputs/{path}_0.pdb\",\n",
        "        f\"--loc=outputs/{path}\",\n",
        "        f\"--contig={contigs_str}\",\n",
        "        f\"--copies={copies}\",\n",
        "        f\"--num_seqs={num_seqs}\",\n",
        "        f\"--num_recycles={num_recycles}\",\n",
        "        f\"--rm_aa={rm_aa}\",\n",
        "        f\"--mpnn_sampling_temp={mpnn_sampling_temp}\",\n",
        "        f\"--num_designs={num_designs}\"]\n",
        "if initial_guess: opts.append(\"--initial_guess\")\n",
        "if use_multimer: opts.append(\"--use_multimer\")\n",
        "opts = ' '.join(opts)\n",
        "!python colabdesign/rf/designability_test.py {opts}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "DUNKRBNSvk6_"
      },
      "outputs": [],
      "source": [
        "#@title Display best result\n",
        "import py3Dmol\n",
        "def plot_pdb(num = \"best\"):\n",
        "  if num == \"best\":\n",
        "    with open(f\"outputs/{path}/best.pdb\",\"r\") as f:\n",
        "      # REMARK 001 design {m} N {n} RMSD {rmsd}\n",
        "      info = f.readline().strip('\\n').split()\n",
        "    num = info[3]\n",
        "  hbondCutoff = 4.0\n",
        "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "  pdb_str = open(f\"outputs/{path}_{num}.pdb\",'r').read()\n",
        "  view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "  pdb_str = open(f\"outputs/{path}/best_design{num}.pdb\",'r').read()\n",
        "  view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "\n",
        "  view.setStyle({\"model\":0},{'cartoon':{}}) #: {'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
        "  view.setStyle({\"model\":1},{'cartoon':{'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
        "  view.zoomTo()\n",
        "  view.show()\n",
        "\n",
        "if num_designs > 1:\n",
        "  def on_change(change):\n",
        "    if change['name'] == 'value':\n",
        "      with output:\n",
        "        output.clear_output(wait=True)\n",
        "        plot_pdb(change['new'])\n",
        "  dropdown = widgets.Dropdown(\n",
        "    options=[\"best\"] + [str(k) for k in range(num_designs)],\n",
        "    value=\"best\",\n",
        "    description='design:',\n",
        "  )\n",
        "  dropdown.observe(on_change)\n",
        "  output = widgets.Output()\n",
        "  display(widgets.VBox([dropdown, output]))\n",
        "  with output:\n",
        "    plot_pdb(dropdown.value)\n",
        "else:\n",
        "  plot_pdb()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "tVAE0BrnZoRR"
      },
      "outputs": [],
      "source": [
        "#@title Package and download results\n",
        "#@markdown If you are having issues downloading the result archive,\n",
        "#@markdown try disabling your adblocker and run this cell again.\n",
        "#@markdown  If that fails click on the little folder icon to the\n",
        "#@markdown  left, navigate to file: `name.result.zip`,\n",
        "#@markdown  right-click and select \\\"Download\\\"\n",
        "#@markdown (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n",
        "!zip -r {path}.result.zip outputs/{path}* outputs/traj/{path}*\n",
        "files.download(f\"{path}.result.zip\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Instructions**\n",
        "---\n",
        "---\n",
        "\n",
        "Use `contigs` to define continious chains. Use a `:` to define multiple contigs and a `/` to define mutliple segments within a contig.\n",
        "For example:\n",
        "\n",
        "**unconditional**\n",
        "- `contigs='100'` - diffuse **monomer** of length 100\n",
        "- `contigs='50:100'` - diffuse **hetero-oligomer** of lengths 50 and 100\n",
        "- `contigs='50'` `symmetry='cyclic'` `order=2` - make two copies of the defined contig(s) and add a symmetry constraint, for **homo-oligomeric** diffusion.\n",
        "\n",
        "**binder design**\n",
        "- `contigs='A:50'` `pdb='4N5T'` - diffuse a **binder** of length 50 to chain A of defined PDB.\n",
        "- `contigs='E6-155:70-100'` `pdb='5KQV'` `hotspot='E64,E88,E96'` - diffuse a **binder** of length 70 to 100 (sampled randomly) to chain E and defined hotspot(s).\n",
        "\n",
        "**motif scaffolding**\n",
        " - `contigs='40/A163-181/40'` `pdb='5TPN'`\n",
        " - `contigs='A3-30/36/A33-68'` `pdb='6MRR'` - diffuse a loop of length 36 between two segments of defined PDB ranges.\n",
        "\n",
        "**partial diffusion**\n",
        "- `contigs=''` `pdb='6MRR'` - noise all coordinates\n",
        "- `contigs='A1-10'` `pdb='6MRR'` - keep first 10 positions fixed, noise the rest\n",
        "- `contigs='A'` `pdb='1SSC'` - fix chain A, noise the rest\n",
        "\n",
        "*hints and tips*\n",
        "- `pdb=''` leave blank to get an upload prompt\n",
        "- `contigs='50-100'` use dash to specify a range of lengths to sample from"
      ],
      "metadata": {
        "id": "DKQXlWEjIOsf"
      }
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}